// @x402-avm/avm/dist/cjs/exact/client/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/exact/client/index.ts
var client_exports = {};
__export(client_exports, {
  ExactAvmScheme: () => ExactAvmScheme,
  registerExactAvmScheme: () => registerExactAvmScheme
});
module.exports = __toCommonJS(client_exports);

// src/exact/client/scheme.ts
var import_transact2 = require("@algorandfoundation/algokit-utils/transact");
var import_common3 = require("@algorandfoundation/algokit-utils/common");

// src/utils.ts
var import_algod_client = require("@algorandfoundation/algokit-utils/algod-client");
var import_transact = require("@algorandfoundation/algokit-utils/transact");
var import_common = require("@algorandfoundation/algokit-utils/common");

// src/constants.ts
var ALGORAND_MAINNET_CAIP2 = "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
var ALGORAND_TESTNET_CAIP2 = "algorand:SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
var V1_ALGORAND_MAINNET = "algorand-mainnet";
var V1_ALGORAND_TESTNET = "algorand-testnet";
var V1_TO_CAIP2 = {
  [V1_ALGORAND_MAINNET]: ALGORAND_MAINNET_CAIP2,
  [V1_ALGORAND_TESTNET]: ALGORAND_TESTNET_CAIP2
};
var CAIP2_TO_V1 = {
  [ALGORAND_MAINNET_CAIP2]: V1_ALGORAND_MAINNET,
  [ALGORAND_TESTNET_CAIP2]: V1_ALGORAND_TESTNET
};
var USDC_MAINNET_ASA_ID = "31566704";
var USDC_TESTNET_ASA_ID = "10458941";
var USDC_DECIMALS = 6;
var USDC_CONFIG = {
  [ALGORAND_MAINNET_CAIP2]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [ALGORAND_TESTNET_CAIP2]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  // V1 network mappings
  [V1_ALGORAND_MAINNET]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [V1_ALGORAND_TESTNET]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  }
};
var FALLBACK_ALGOD_MAINNET = "https://mainnet-api.algonode.cloud";
var FALLBACK_ALGOD_TESTNET = "https://testnet-api.algonode.cloud";
var DEFAULT_ALGOD_MAINNET = typeof process !== "undefined" && process.env?.ALGOD_MAINNET_URL || FALLBACK_ALGOD_MAINNET;
var DEFAULT_ALGOD_TESTNET = typeof process !== "undefined" && process.env?.ALGOD_TESTNET_URL || FALLBACK_ALGOD_TESTNET;
var NETWORK_TO_ALGOD = {
  [ALGORAND_MAINNET_CAIP2]: DEFAULT_ALGOD_MAINNET,
  [ALGORAND_TESTNET_CAIP2]: DEFAULT_ALGOD_TESTNET,
  [V1_ALGORAND_MAINNET]: DEFAULT_ALGOD_MAINNET,
  [V1_ALGORAND_TESTNET]: DEFAULT_ALGOD_TESTNET
};

// src/utils.ts
var import_algod_client2 = require("@algorandfoundation/algokit-utils/algod-client");
var import_common2 = require("@algorandfoundation/algokit-utils/common");
function createAlgodClient(network, customUrl, token = "") {
  const url = customUrl ?? NETWORK_TO_ALGOD[network] ?? DEFAULT_ALGOD_TESTNET;
  return new import_algod_client.AlgodClient({ baseUrl: url, token: token || void 0 });
}
function encodeTransaction(txn) {
  return Buffer.from(txn).toString("base64");
}

// src/exact/client/scheme.ts
var ExactAvmScheme = class {
  /**
   * Creates a new ExactAvmScheme instance.
   *
   * @param signer - The AVM signer for client operations
   * @param config - Optional configuration for Algod client
   */
  constructor(signer, config) {
    this.signer = signer;
    this.config = config;
    this.scheme = "exact";
  }
  /**
   * Creates a payment payload for the Exact scheme.
   *
   * Constructs an atomic transaction group with:
   * - Optional fee payer transaction (if feePayer specified in requirements.extra)
   * - ASA transfer transaction to payTo address
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements
   * @returns Promise resolving to a payment payload result
   */
  async createPaymentPayload(x402Version, paymentRequirements) {
    const { amount, asset, payTo, network, extra } = paymentRequirements;
    const algodClient = this.config?.algodClient ?? createAlgodClient(
      network,
      this.config?.algodUrl ?? DEFAULT_ALGOD_TESTNET,
      this.config?.algodToken
    );
    const suggestedParams = await algodClient.suggestedParams();
    const assetId = this.getAssetId(asset, network);
    const feePayer = extra?.feePayer;
    const transactions = [];
    let paymentIndex = 0;
    const totalTxnCount = feePayer ? 2 : 1;
    const minFee = suggestedParams.minFee ?? BigInt(1e3);
    if (feePayer) {
      const feePayerTxn = new import_transact2.Transaction({
        type: import_transact2.TransactionType.Payment,
        sender: import_common3.Address.fromString(feePayer),
        fee: minFee * BigInt(totalTxnCount),
        firstValid: suggestedParams.firstValid,
        lastValid: suggestedParams.lastValid,
        genesisHash: suggestedParams.genesisHash,
        genesisId: suggestedParams.genesisId,
        note: new Uint8Array(Buffer.from(`x402-fee-payer-${Date.now()}`)),
        payment: {
          receiver: import_common3.Address.fromString(feePayer),
          amount: BigInt(0)
        }
      });
      transactions.push(feePayerTxn);
      paymentIndex = 1;
    }
    const assetTransferFee = feePayer ? BigInt(0) : suggestedParams.fee ?? minFee;
    const assetTransferTxn = new import_transact2.Transaction({
      type: import_transact2.TransactionType.AssetTransfer,
      sender: import_common3.Address.fromString(this.signer.address),
      fee: assetTransferFee,
      firstValid: suggestedParams.firstValid,
      lastValid: suggestedParams.lastValid,
      genesisHash: suggestedParams.genesisHash,
      genesisId: suggestedParams.genesisId,
      note: new Uint8Array(Buffer.from(`x402-payment-v${x402Version}-${Date.now()}`)),
      assetTransfer: {
        receiver: import_common3.Address.fromString(payTo),
        amount: BigInt(amount),
        assetId: BigInt(assetId)
      }
    });
    transactions.push(assetTransferTxn);
    let groupedTxns = transactions;
    if (transactions.length > 1) {
      groupedTxns = (0, import_transact2.groupTransactions)(transactions);
    }
    const encodedTxns = groupedTxns.map((txn) => (0, import_transact2.encodeTransactionRaw)(txn));
    const clientIndexes = groupedTxns.map((txn, i) => {
      const sender = txn.sender.toString();
      return sender === this.signer.address ? i : -1;
    }).filter((i) => i !== -1);
    console.log("[x402 AVM Client] Creating payment:", {
      sender: this.signer.address,
      receiver: payTo,
      amount,
      assetId: this.getAssetId(asset, network),
      network,
      clientIndexes,
      txnCount: groupedTxns.length,
      hasFeePayer: !!feePayer
    });
    const signedTxns = await this.signer.signTransactions(encodedTxns, clientIndexes);
    console.log("[x402 AVM Client] Signed transactions:", {
      signedCount: signedTxns.filter((t) => t !== null).length,
      totalCount: signedTxns.length,
      signedIndexes: signedTxns.map((t, i) => t !== null ? i : -1).filter((i) => i !== -1)
    });
    const paymentGroup = encodedTxns.map((txnBytes, i) => {
      const signedTxn = signedTxns[i];
      if (signedTxn) {
        return encodeTransaction(signedTxn);
      }
      return encodeTransaction(txnBytes);
    });
    const payload = {
      paymentGroup,
      paymentIndex
    };
    return {
      x402Version,
      payload
    };
  }
  /**
   * Gets the asset ID from the requirements or defaults to USDC
   *
   * @param asset - Asset identifier from requirements
   * @param network - Network identifier
   * @returns Asset ID as string
   */
  getAssetId(asset, network) {
    if (/^\d+$/.test(asset)) {
      return asset;
    }
    const usdcConfig = USDC_CONFIG[network];
    if (usdcConfig) {
      return usdcConfig.asaId;
    }
    return asset;
  }
};

// src/exact/v1/client/scheme.ts
var import_transact3 = require("@algorandfoundation/algokit-utils/transact");
var import_common4 = require("@algorandfoundation/algokit-utils/common");
var ExactAvmSchemeV1 = class {
  /**
   * Creates a new ExactAvmSchemeV1 instance.
   *
   * @param signer - The AVM signer for client operations
   * @param config - Optional configuration for Algod client
   */
  constructor(signer, config) {
    this.signer = signer;
    this.config = config;
    this.scheme = "exact";
  }
  /**
   * Creates a payment payload for the Exact scheme (V1).
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements
   * @returns Promise resolving to a payment payload
   */
  async createPaymentPayload(x402Version, paymentRequirements) {
    const selectedV1 = paymentRequirements;
    const v1Network = selectedV1.network;
    const caip2Network = V1_TO_CAIP2[v1Network] ?? v1Network;
    const algodClient = this.config?.algodClient ?? createAlgodClient(
      caip2Network,
      this.config?.algodUrl ?? DEFAULT_ALGOD_TESTNET,
      this.config?.algodToken
    );
    const suggestedParams = await algodClient.suggestedParams();
    const assetId = this.getAssetId(selectedV1.asset, caip2Network);
    const assetTransferTxn = new import_transact3.Transaction({
      type: import_transact3.TransactionType.AssetTransfer,
      sender: import_common4.Address.fromString(this.signer.address),
      fee: suggestedParams.fee ?? suggestedParams.minFee ?? BigInt(1e3),
      firstValid: suggestedParams.firstValid,
      lastValid: suggestedParams.lastValid,
      genesisHash: suggestedParams.genesisHash,
      genesisId: suggestedParams.genesisId,
      note: new Uint8Array(Buffer.from(`x402-payment-v${x402Version}`)),
      assetTransfer: {
        receiver: import_common4.Address.fromString(selectedV1.payTo),
        amount: BigInt(selectedV1.maxAmountRequired),
        assetId: BigInt(assetId)
      }
    });
    const encodedTxn = (0, import_transact3.encodeTransactionRaw)(assetTransferTxn);
    const signedTxns = await this.signer.signTransactions([encodedTxn], [0]);
    const signedTxn = signedTxns[0];
    if (!signedTxn) {
      throw new Error("Failed to sign transaction");
    }
    const payload = {
      paymentGroup: [encodeTransaction(signedTxn)],
      paymentIndex: 0
    };
    return {
      x402Version,
      scheme: selectedV1.scheme,
      network: selectedV1.network,
      payload
    };
  }
  /**
   * Gets the asset ID from the requirements or defaults to USDC
   *
   * @param asset - The asset identifier from requirements
   * @param network - The network identifier
   * @returns The resolved ASA ID string
   */
  getAssetId(asset, network) {
    if (/^\d+$/.test(asset)) {
      return asset;
    }
    const usdcConfig = USDC_CONFIG[network];
    if (usdcConfig) {
      return usdcConfig.asaId;
    }
    return asset;
  }
};

// src/v1/index.ts
var AVM_NETWORK_CAIP2_MAP = {
  "algorand-mainnet": "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=",
  "algorand-testnet": "algorand:SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI="
};
var NETWORKS = Object.keys(AVM_NETWORK_CAIP2_MAP);

// src/exact/client/register.ts
function registerExactAvmScheme(client, config) {
  const scheme = new ExactAvmScheme(config.signer, config.algodConfig);
  if (config.networks && config.networks.length > 0) {
    config.networks.forEach((network) => {
      client.register(network, scheme);
    });
  } else {
    client.register("algorand:*", scheme);
  }
  const v1Scheme = new ExactAvmSchemeV1(config.signer, config.algodConfig);
  NETWORKS.forEach((network) => {
    client.registerV1(network, v1Scheme);
  });
  if (config.policies) {
    config.policies.forEach((policy) => {
      client.registerPolicy(policy);
    });
  }
  return client;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ExactAvmScheme,
  registerExactAvmScheme
});
