// @x402-avm/avm/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  ALGORAND_ADDRESS_LENGTH: () => ALGORAND_ADDRESS_LENGTH,
  ALGORAND_ADDRESS_REGEX: () => ALGORAND_ADDRESS_REGEX,
  ALGORAND_MAINNET_CAIP2: () => ALGORAND_MAINNET_CAIP2,
  ALGORAND_MAINNET_GENESIS_HASH: () => ALGORAND_MAINNET_GENESIS_HASH,
  ALGORAND_TESTNET_CAIP2: () => ALGORAND_TESTNET_CAIP2,
  ALGORAND_TESTNET_GENESIS_HASH: () => ALGORAND_TESTNET_GENESIS_HASH,
  CAIP2_NETWORKS: () => CAIP2_NETWORKS,
  CAIP2_TO_V1: () => CAIP2_TO_V1,
  DEFAULT_ALGOD_MAINNET: () => DEFAULT_ALGOD_MAINNET,
  DEFAULT_ALGOD_TESTNET: () => DEFAULT_ALGOD_TESTNET,
  ExactAvmScheme: () => ExactAvmScheme,
  MAX_ATOMIC_GROUP_SIZE: () => MAX_ATOMIC_GROUP_SIZE,
  MAX_REASONABLE_FEE: () => MAX_REASONABLE_FEE,
  MIN_TXN_FEE: () => MIN_TXN_FEE,
  NETWORK_TO_ALGOD: () => NETWORK_TO_ALGOD,
  USDC_CONFIG: () => USDC_CONFIG,
  USDC_DECIMALS: () => USDC_DECIMALS,
  USDC_MAINNET_ASA_ID: () => USDC_MAINNET_ASA_ID,
  USDC_TESTNET_ASA_ID: () => USDC_TESTNET_ASA_ID,
  V1_ALGORAND_MAINNET: () => V1_ALGORAND_MAINNET,
  V1_ALGORAND_TESTNET: () => V1_ALGORAND_TESTNET,
  V1_NETWORKS: () => V1_NETWORKS,
  V1_TO_CAIP2: () => V1_TO_CAIP2,
  assignGroupId: () => assignGroupId,
  caip2ToV1: () => caip2ToV1,
  convertFromTokenAmount: () => convertFromTokenAmount,
  convertToTokenAmount: () => convertToTokenAmount,
  createAlgodClient: () => createAlgodClient,
  decodeSignedTransaction: () => decodeSignedTransaction,
  decodeTransaction: () => decodeTransaction,
  decodeUnsignedTransaction: () => decodeUnsignedTransaction,
  encodeTransaction: () => encodeTransaction,
  getGenesisHashFromTransaction: () => getGenesisHashFromTransaction,
  getNetworkFromCaip2: () => getNetworkFromCaip2,
  getSenderFromTransaction: () => getSenderFromTransaction,
  getTransactionId: () => getTransactionId,
  hasSignature: () => hasSignature,
  isAlgorandNetwork: () => isAlgorandNetwork,
  isAvmSignerWallet: () => isAvmSignerWallet,
  isExactAvmPayload: () => isExactAvmPayload,
  isTestnetNetwork: () => isTestnetNetwork,
  isValidAlgorandAddress: () => isValidAlgorandAddress,
  toClientAvmSigner: () => toClientAvmSigner,
  toFacilitatorAvmSigner: () => toFacilitatorAvmSigner,
  v1ToCaip2: () => v1ToCaip2,
  validateGroupId: () => validateGroupId
});
module.exports = __toCommonJS(src_exports);

// src/exact/client/scheme.ts
var import_transact2 = require("@algorandfoundation/algokit-utils/transact");
var import_common3 = require("@algorandfoundation/algokit-utils/common");

// src/utils.ts
var import_algod_client = require("@algorandfoundation/algokit-utils/algod-client");
var import_transact = require("@algorandfoundation/algokit-utils/transact");
var import_common = require("@algorandfoundation/algokit-utils/common");

// src/constants.ts
var ALGORAND_MAINNET_CAIP2 = "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
var ALGORAND_TESTNET_CAIP2 = "algorand:SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
var CAIP2_NETWORKS = [ALGORAND_MAINNET_CAIP2, ALGORAND_TESTNET_CAIP2];
var ALGORAND_MAINNET_GENESIS_HASH = "wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
var ALGORAND_TESTNET_GENESIS_HASH = "SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
var V1_ALGORAND_MAINNET = "algorand-mainnet";
var V1_ALGORAND_TESTNET = "algorand-testnet";
var V1_NETWORKS = [V1_ALGORAND_MAINNET, V1_ALGORAND_TESTNET];
var V1_TO_CAIP2 = {
  [V1_ALGORAND_MAINNET]: ALGORAND_MAINNET_CAIP2,
  [V1_ALGORAND_TESTNET]: ALGORAND_TESTNET_CAIP2
};
var CAIP2_TO_V1 = {
  [ALGORAND_MAINNET_CAIP2]: V1_ALGORAND_MAINNET,
  [ALGORAND_TESTNET_CAIP2]: V1_ALGORAND_TESTNET
};
var USDC_MAINNET_ASA_ID = "31566704";
var USDC_TESTNET_ASA_ID = "10458941";
var USDC_DECIMALS = 6;
var USDC_CONFIG = {
  [ALGORAND_MAINNET_CAIP2]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [ALGORAND_TESTNET_CAIP2]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  // V1 network mappings
  [V1_ALGORAND_MAINNET]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [V1_ALGORAND_TESTNET]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  }
};
var FALLBACK_ALGOD_MAINNET = "https://mainnet-api.algonode.cloud";
var FALLBACK_ALGOD_TESTNET = "https://testnet-api.algonode.cloud";
var DEFAULT_ALGOD_MAINNET = typeof process !== "undefined" && process.env?.ALGOD_MAINNET_URL || FALLBACK_ALGOD_MAINNET;
var DEFAULT_ALGOD_TESTNET = typeof process !== "undefined" && process.env?.ALGOD_TESTNET_URL || FALLBACK_ALGOD_TESTNET;
var NETWORK_TO_ALGOD = {
  [ALGORAND_MAINNET_CAIP2]: DEFAULT_ALGOD_MAINNET,
  [ALGORAND_TESTNET_CAIP2]: DEFAULT_ALGOD_TESTNET,
  [V1_ALGORAND_MAINNET]: DEFAULT_ALGOD_MAINNET,
  [V1_ALGORAND_TESTNET]: DEFAULT_ALGOD_TESTNET
};
var MAX_ATOMIC_GROUP_SIZE = 16;
var MIN_TXN_FEE = 1e3;
var MAX_REASONABLE_FEE = 16e3;
var ALGORAND_ADDRESS_REGEX = /^[A-Z2-7]{58}$/;
var ALGORAND_ADDRESS_LENGTH = 58;

// src/utils.ts
var import_algod_client2 = require("@algorandfoundation/algokit-utils/algod-client");
var import_common2 = require("@algorandfoundation/algokit-utils/common");
function createAlgodClient(network, customUrl, token = "") {
  const url = customUrl ?? NETWORK_TO_ALGOD[network] ?? DEFAULT_ALGOD_TESTNET;
  return new import_algod_client.AlgodClient({ baseUrl: url, token: token || void 0 });
}
function encodeTransaction(txn) {
  return Buffer.from(txn).toString("base64");
}
function decodeTransaction(encoded) {
  return new Uint8Array(Buffer.from(encoded, "base64"));
}
function decodeSignedTransaction(encoded) {
  const bytes = decodeTransaction(encoded);
  return (0, import_transact.decodeSignedTransaction)(bytes);
}
function decodeUnsignedTransaction(encoded) {
  const bytes = decodeTransaction(encoded);
  return (0, import_transact.decodeTransaction)(bytes);
}
function isValidAlgorandAddress(address) {
  if (!ALGORAND_ADDRESS_REGEX.test(address)) {
    return false;
  }
  return (0, import_common.isValidAddress)(address);
}
function getSenderFromTransaction(txnBytes, isSigned = true) {
  if (isSigned) {
    const signedTxn = (0, import_transact.decodeSignedTransaction)(txnBytes);
    return signedTxn.txn.sender.toString();
  }
  const txn = (0, import_transact.decodeTransaction)(txnBytes);
  return txn.sender.toString();
}
function convertToTokenAmount(decimalAmount, decimals) {
  const amount = parseFloat(decimalAmount);
  if (isNaN(amount)) {
    throw new Error(`Invalid amount: ${decimalAmount}`);
  }
  const [intPart, decPart = ""] = String(amount).split(".");
  const paddedDec = decPart.padEnd(decimals, "0").slice(0, decimals);
  const tokenAmount = (intPart + paddedDec).replace(/^0+/, "") || "0";
  return tokenAmount;
}
function convertFromTokenAmount(atomicAmount, decimals) {
  const amount = BigInt(atomicAmount);
  const divisor = BigInt(10 ** decimals);
  const intPart = amount / divisor;
  const decPart = amount % divisor;
  if (decPart === BigInt(0)) {
    return intPart.toString();
  }
  const decStr = decPart.toString().padStart(decimals, "0");
  const trimmedDec = decStr.replace(/0+$/, "");
  return `${intPart}.${trimmedDec}`;
}
function getNetworkFromCaip2(caip2) {
  if (!caip2.startsWith("algorand:")) {
    return null;
  }
  const genesisHash = caip2.slice("algorand:".length);
  if (genesisHash === ALGORAND_MAINNET_GENESIS_HASH) {
    return "mainnet";
  }
  if (genesisHash === ALGORAND_TESTNET_GENESIS_HASH) {
    return "testnet";
  }
  return null;
}
function isAlgorandNetwork(network) {
  if (network.startsWith("algorand:")) {
    return true;
  }
  return network === V1_ALGORAND_MAINNET || network === V1_ALGORAND_TESTNET;
}
function isTestnetNetwork(network) {
  return network === ALGORAND_TESTNET_CAIP2 || network === V1_ALGORAND_TESTNET;
}
function v1ToCaip2(v1Network) {
  return V1_TO_CAIP2[v1Network] ?? v1Network;
}
function caip2ToV1(caip2Network) {
  return CAIP2_TO_V1[caip2Network] ?? caip2Network;
}
function getGenesisHashFromTransaction(txn) {
  if (!txn.genesisHash) {
    throw new Error("Transaction does not have a genesis hash");
  }
  return Buffer.from(txn.genesisHash).toString("base64");
}
function validateGroupId(txns) {
  if (txns.length <= 1) {
    return true;
  }
  let expectedGroupId = null;
  for (const txnBytes of txns) {
    const txn = (0, import_transact.decodeTransaction)(txnBytes);
    const groupId = txn.group ? Buffer.from(txn.group).toString("base64") : null;
    if (expectedGroupId === null) {
      expectedGroupId = groupId;
    } else if (groupId !== expectedGroupId) {
      return false;
    }
  }
  return true;
}
function assignGroupId(txns) {
  if (txns.length <= 1) {
    return txns;
  }
  return (0, import_transact.groupTransactions)(txns);
}
function getTransactionId(signedTxnBytes) {
  const signedTxn = (0, import_transact.decodeSignedTransaction)(signedTxnBytes);
  return signedTxn.txn.txId();
}
function hasSignature(signedTxnBytes) {
  const signedTxn = (0, import_transact.decodeSignedTransaction)(signedTxnBytes);
  return signedTxn.sig !== void 0 || signedTxn.lsig !== void 0 || signedTxn.msig !== void 0;
}

// src/exact/client/scheme.ts
var ExactAvmScheme = class {
  /**
   * Creates a new ExactAvmScheme instance.
   *
   * @param signer - The AVM signer for client operations
   * @param config - Optional configuration for Algod client
   */
  constructor(signer, config) {
    this.signer = signer;
    this.config = config;
    this.scheme = "exact";
  }
  /**
   * Creates a payment payload for the Exact scheme.
   *
   * Constructs an atomic transaction group with:
   * - Optional fee payer transaction (if feePayer specified in requirements.extra)
   * - ASA transfer transaction to payTo address
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements
   * @returns Promise resolving to a payment payload result
   */
  async createPaymentPayload(x402Version, paymentRequirements) {
    const { amount, asset, payTo, network, extra } = paymentRequirements;
    const algodClient = this.config?.algodClient ?? createAlgodClient(
      network,
      this.config?.algodUrl ?? DEFAULT_ALGOD_TESTNET,
      this.config?.algodToken
    );
    const suggestedParams = await algodClient.suggestedParams();
    const assetId = this.getAssetId(asset, network);
    const feePayer = extra?.feePayer;
    const transactions = [];
    let paymentIndex = 0;
    const totalTxnCount = feePayer ? 2 : 1;
    const minFee = suggestedParams.minFee ?? BigInt(1e3);
    if (feePayer) {
      const feePayerTxn = new import_transact2.Transaction({
        type: import_transact2.TransactionType.Payment,
        sender: import_common3.Address.fromString(feePayer),
        fee: minFee * BigInt(totalTxnCount),
        firstValid: suggestedParams.firstValid,
        lastValid: suggestedParams.lastValid,
        genesisHash: suggestedParams.genesisHash,
        genesisId: suggestedParams.genesisId,
        note: new Uint8Array(Buffer.from(`x402-fee-payer-${Date.now()}`)),
        payment: {
          receiver: import_common3.Address.fromString(feePayer),
          amount: BigInt(0)
        }
      });
      transactions.push(feePayerTxn);
      paymentIndex = 1;
    }
    const assetTransferFee = feePayer ? BigInt(0) : suggestedParams.fee ?? minFee;
    const assetTransferTxn = new import_transact2.Transaction({
      type: import_transact2.TransactionType.AssetTransfer,
      sender: import_common3.Address.fromString(this.signer.address),
      fee: assetTransferFee,
      firstValid: suggestedParams.firstValid,
      lastValid: suggestedParams.lastValid,
      genesisHash: suggestedParams.genesisHash,
      genesisId: suggestedParams.genesisId,
      note: new Uint8Array(Buffer.from(`x402-payment-v${x402Version}-${Date.now()}`)),
      assetTransfer: {
        receiver: import_common3.Address.fromString(payTo),
        amount: BigInt(amount),
        assetId: BigInt(assetId)
      }
    });
    transactions.push(assetTransferTxn);
    let groupedTxns = transactions;
    if (transactions.length > 1) {
      groupedTxns = (0, import_transact2.groupTransactions)(transactions);
    }
    const encodedTxns = groupedTxns.map((txn) => (0, import_transact2.encodeTransactionRaw)(txn));
    const clientIndexes = groupedTxns.map((txn, i) => {
      const sender = txn.sender.toString();
      return sender === this.signer.address ? i : -1;
    }).filter((i) => i !== -1);
    console.log("[x402 AVM Client] Creating payment:", {
      sender: this.signer.address,
      receiver: payTo,
      amount,
      assetId: this.getAssetId(asset, network),
      network,
      clientIndexes,
      txnCount: groupedTxns.length,
      hasFeePayer: !!feePayer
    });
    const signedTxns = await this.signer.signTransactions(encodedTxns, clientIndexes);
    console.log("[x402 AVM Client] Signed transactions:", {
      signedCount: signedTxns.filter((t) => t !== null).length,
      totalCount: signedTxns.length,
      signedIndexes: signedTxns.map((t, i) => t !== null ? i : -1).filter((i) => i !== -1)
    });
    const paymentGroup = encodedTxns.map((txnBytes, i) => {
      const signedTxn = signedTxns[i];
      if (signedTxn) {
        return encodeTransaction(signedTxn);
      }
      return encodeTransaction(txnBytes);
    });
    const payload = {
      paymentGroup,
      paymentIndex
    };
    return {
      x402Version,
      payload
    };
  }
  /**
   * Gets the asset ID from the requirements or defaults to USDC
   *
   * @param asset - Asset identifier from requirements
   * @param network - Network identifier
   * @returns Asset ID as string
   */
  getAssetId(asset, network) {
    if (/^\d+$/.test(asset)) {
      return asset;
    }
    const usdcConfig = USDC_CONFIG[network];
    if (usdcConfig) {
      return usdcConfig.asaId;
    }
    return asset;
  }
};

// src/signer.ts
var import_algorand_client = require("@algorandfoundation/algokit-utils/algorand-client");
var import_common4 = require("@algorandfoundation/algokit-utils/common");
var import_crypto = require("@algorandfoundation/algokit-utils/crypto");
var import_transact3 = require("@algorandfoundation/algokit-utils/transact");
var import_transaction = require("@algorandfoundation/algokit-utils/transaction");
function isAvmSignerWallet(wallet) {
  return typeof wallet === "object" && wallet !== null && "address" in wallet && typeof wallet.address === "string" && "signTransactions" in wallet && typeof wallet.signTransactions === "function";
}
function decodePrivateKey(privateKeyBase64) {
  const secretKey = Buffer.from(privateKeyBase64, "base64");
  if (secretKey.length !== 64) {
    throw new Error(
      "AVM private key must be a Base64-encoded 64-byte key (32-byte seed + 32-byte public key)"
    );
  }
  const seed = secretKey.slice(0, 32);
  const { ed25519Pubkey, rawEd25519Signer } = (0, import_crypto.ed25519Generator)(seed);
  const address = (0, import_common4.encodeAddress)(ed25519Pubkey);
  return { address, rawEd25519Signer };
}
function toClientAvmSigner(privateKeyBase64) {
  const { address, rawEd25519Signer } = decodePrivateKey(privateKeyBase64);
  return {
    address,
    signTransactions: async (txns, indexesToSign) => {
      return Promise.all(
        txns.map(async (txn, i) => {
          if (indexesToSign && !indexesToSign.includes(i)) return null;
          const decoded = (0, import_transact3.decodeTransaction)(txn);
          const msg = import_transact3.bytesForSigning.transaction(decoded);
          const sig = await rawEd25519Signer(msg);
          return (0, import_transact3.encodeSignedTransaction)({ txn: decoded, sig });
        })
      );
    }
  };
}
function isTestnet(network) {
  return network === ALGORAND_TESTNET_CAIP2 || network === V1_ALGORAND_TESTNET;
}
function toFacilitatorAvmSigner(privateKeyBase64, config) {
  const { address, rawEd25519Signer } = decodePrivateKey(privateKeyBase64);
  const getAlgorandClientForNetwork = (network) => {
    if (isTestnet(network)) {
      if (config?.testnetUrl) {
        return import_algorand_client.AlgorandClient.fromConfig({
          algodConfig: { server: config.testnetUrl, token: config.algodToken ?? "" }
        });
      }
      return import_algorand_client.AlgorandClient.testNet();
    }
    if (config?.mainnetUrl) {
      return import_algorand_client.AlgorandClient.fromConfig({
        algodConfig: { server: config.mainnetUrl, token: config.algodToken ?? "" }
      });
    }
    return import_algorand_client.AlgorandClient.mainNet();
  };
  const clientCache = /* @__PURE__ */ new Map();
  const getClient = (network) => {
    const key = isTestnet(network) ? "testnet" : "mainnet";
    let client = clientCache.get(key);
    if (!client) {
      client = getAlgorandClientForNetwork(network);
      clientCache.set(key, client);
    }
    return client;
  };
  return {
    getAddresses: () => [address],
    signTransaction: async (txn, _) => {
      const decoded = (0, import_transact3.decodeTransaction)(txn);
      const msg = import_transact3.bytesForSigning.transaction(decoded);
      const sig = await rawEd25519Signer(msg);
      return (0, import_transact3.encodeSignedTransaction)({ txn: decoded, sig });
    },
    getAlgodClient: (network) => getClient(network).client.algod,
    simulateTransactions: async (txns, network) => {
      const algod = getClient(network).client.algod;
      return await algod.simulateRawTransactions(txns);
    },
    sendTransactions: async (signedTxns, network) => {
      const algod = getClient(network).client.algod;
      const response = await algod.sendRawTransaction(signedTxns);
      return response.txId;
    },
    waitForConfirmation: async (txId, network, waitRounds = 4) => {
      const algod = getClient(network).client.algod;
      return await (0, import_transaction.waitForConfirmation)(txId, waitRounds, algod);
    }
  };
}

// src/types.ts
function isExactAvmPayload(payload) {
  return typeof payload === "object" && payload !== null && "paymentGroup" in payload && "paymentIndex" in payload && Array.isArray(payload.paymentGroup) && typeof payload.paymentIndex === "number" && payload.paymentGroup.every((item) => typeof item === "string");
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ALGORAND_ADDRESS_LENGTH,
  ALGORAND_ADDRESS_REGEX,
  ALGORAND_MAINNET_CAIP2,
  ALGORAND_MAINNET_GENESIS_HASH,
  ALGORAND_TESTNET_CAIP2,
  ALGORAND_TESTNET_GENESIS_HASH,
  CAIP2_NETWORKS,
  CAIP2_TO_V1,
  DEFAULT_ALGOD_MAINNET,
  DEFAULT_ALGOD_TESTNET,
  ExactAvmScheme,
  MAX_ATOMIC_GROUP_SIZE,
  MAX_REASONABLE_FEE,
  MIN_TXN_FEE,
  NETWORK_TO_ALGOD,
  USDC_CONFIG,
  USDC_DECIMALS,
  USDC_MAINNET_ASA_ID,
  USDC_TESTNET_ASA_ID,
  V1_ALGORAND_MAINNET,
  V1_ALGORAND_TESTNET,
  V1_NETWORKS,
  V1_TO_CAIP2,
  assignGroupId,
  caip2ToV1,
  convertFromTokenAmount,
  convertToTokenAmount,
  createAlgodClient,
  decodeSignedTransaction,
  decodeTransaction,
  decodeUnsignedTransaction,
  encodeTransaction,
  getGenesisHashFromTransaction,
  getNetworkFromCaip2,
  getSenderFromTransaction,
  getTransactionId,
  hasSignature,
  isAlgorandNetwork,
  isAvmSignerWallet,
  isExactAvmPayload,
  isTestnetNetwork,
  isValidAlgorandAddress,
  toClientAvmSigner,
  toFacilitatorAvmSigner,
  v1ToCaip2,
  validateGroupId
});
