// @x402-avm/avm/dist/esm/chunk-EXTKAZHL.mjs
import {
  createAlgodClient,
  encodeTransaction
} from "./chunk-TLHHCNYP.mjs";
import {
  DEFAULT_ALGOD_TESTNET,
  USDC_CONFIG,
  V1_TO_CAIP2
} from "./chunk-DUGAE7GK.mjs";

// src/exact/v1/client/scheme.ts
import {
  Transaction,
  TransactionType,
  encodeTransactionRaw
} from "@algorandfoundation/algokit-utils/transact";
import { Address } from "@algorandfoundation/algokit-utils/common";
var ExactAvmSchemeV1 = class {
  /**
   * Creates a new ExactAvmSchemeV1 instance.
   *
   * @param signer - The AVM signer for client operations
   * @param config - Optional configuration for Algod client
   */
  constructor(signer, config) {
    this.signer = signer;
    this.config = config;
    this.scheme = "exact";
  }
  /**
   * Creates a payment payload for the Exact scheme (V1).
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements
   * @returns Promise resolving to a payment payload
   */
  async createPaymentPayload(x402Version, paymentRequirements) {
    const selectedV1 = paymentRequirements;
    const v1Network = selectedV1.network;
    const caip2Network = V1_TO_CAIP2[v1Network] ?? v1Network;
    const algodClient = this.config?.algodClient ?? createAlgodClient(
      caip2Network,
      this.config?.algodUrl ?? DEFAULT_ALGOD_TESTNET,
      this.config?.algodToken
    );
    const suggestedParams = await algodClient.suggestedParams();
    const assetId = this.getAssetId(selectedV1.asset, caip2Network);
    const assetTransferTxn = new Transaction({
      type: TransactionType.AssetTransfer,
      sender: Address.fromString(this.signer.address),
      fee: suggestedParams.fee ?? suggestedParams.minFee ?? BigInt(1e3),
      firstValid: suggestedParams.firstValid,
      lastValid: suggestedParams.lastValid,
      genesisHash: suggestedParams.genesisHash,
      genesisId: suggestedParams.genesisId,
      note: new Uint8Array(Buffer.from(`x402-payment-v${x402Version}`)),
      assetTransfer: {
        receiver: Address.fromString(selectedV1.payTo),
        amount: BigInt(selectedV1.maxAmountRequired),
        assetId: BigInt(assetId)
      }
    });
    const encodedTxn = encodeTransactionRaw(assetTransferTxn);
    const signedTxns = await this.signer.signTransactions([encodedTxn], [0]);
    const signedTxn = signedTxns[0];
    if (!signedTxn) {
      throw new Error("Failed to sign transaction");
    }
    const payload = {
      paymentGroup: [encodeTransaction(signedTxn)],
      paymentIndex: 0
    };
    return {
      x402Version,
      scheme: selectedV1.scheme,
      network: selectedV1.network,
      payload
    };
  }
  /**
   * Gets the asset ID from the requirements or defaults to USDC
   *
   * @param asset - The asset identifier from requirements
   * @param network - The network identifier
   * @returns The resolved ASA ID string
   */
  getAssetId(asset, network) {
    if (/^\d+$/.test(asset)) {
      return asset;
    }
    const usdcConfig = USDC_CONFIG[network];
    if (usdcConfig) {
      return usdcConfig.asaId;
    }
    return asset;
  }
};

export {
  ExactAvmSchemeV1
};
