// @x402-avm/avm/dist/esm/index.mjs
import {
  ExactAvmScheme
} from "./chunk-ZXLI3SMZ.mjs";
import {
  isExactAvmPayload
} from "./chunk-IHTCCPJQ.mjs";
import {
  assignGroupId,
  caip2ToV1,
  convertFromTokenAmount,
  convertToTokenAmount,
  createAlgodClient,
  decodeSignedTransaction,
  decodeTransaction,
  decodeUnsignedTransaction,
  encodeTransaction,
  getGenesisHashFromTransaction,
  getNetworkFromCaip2,
  getSenderFromTransaction,
  getTransactionId,
  hasSignature,
  isAlgorandNetwork,
  isTestnetNetwork,
  isValidAlgorandAddress,
  v1ToCaip2,
  validateGroupId
} from "./chunk-TLHHCNYP.mjs";
import {
  ALGORAND_ADDRESS_LENGTH,
  ALGORAND_ADDRESS_REGEX,
  ALGORAND_MAINNET_CAIP2,
  ALGORAND_MAINNET_GENESIS_HASH,
  ALGORAND_TESTNET_CAIP2,
  ALGORAND_TESTNET_GENESIS_HASH,
  CAIP2_NETWORKS,
  CAIP2_TO_V1,
  DEFAULT_ALGOD_MAINNET,
  DEFAULT_ALGOD_TESTNET,
  MAX_ATOMIC_GROUP_SIZE,
  MAX_REASONABLE_FEE,
  MIN_TXN_FEE,
  NETWORK_TO_ALGOD,
  USDC_CONFIG,
  USDC_DECIMALS,
  USDC_MAINNET_ASA_ID,
  USDC_TESTNET_ASA_ID,
  V1_ALGORAND_MAINNET,
  V1_ALGORAND_TESTNET,
  V1_NETWORKS,
  V1_TO_CAIP2
} from "./chunk-DUGAE7GK.mjs";

// src/signer.ts
import { AlgorandClient } from "@algorandfoundation/algokit-utils/algorand-client";
import { encodeAddress } from "@algorandfoundation/algokit-utils/common";
import { ed25519Generator } from "@algorandfoundation/algokit-utils/crypto";
import {
  decodeTransaction as decodeTransaction2,
  bytesForSigning,
  encodeSignedTransaction
} from "@algorandfoundation/algokit-utils/transact";
import { waitForConfirmation } from "@algorandfoundation/algokit-utils/transaction";
function isAvmSignerWallet(wallet) {
  return typeof wallet === "object" && wallet !== null && "address" in wallet && typeof wallet.address === "string" && "signTransactions" in wallet && typeof wallet.signTransactions === "function";
}
function decodePrivateKey(privateKeyBase64) {
  const secretKey = Buffer.from(privateKeyBase64, "base64");
  if (secretKey.length !== 64) {
    throw new Error(
      "AVM private key must be a Base64-encoded 64-byte key (32-byte seed + 32-byte public key)"
    );
  }
  const seed = secretKey.slice(0, 32);
  const { ed25519Pubkey, rawEd25519Signer } = ed25519Generator(seed);
  const address = encodeAddress(ed25519Pubkey);
  return { address, rawEd25519Signer };
}
function toClientAvmSigner(privateKeyBase64) {
  const { address, rawEd25519Signer } = decodePrivateKey(privateKeyBase64);
  return {
    address,
    signTransactions: async (txns, indexesToSign) => {
      return Promise.all(
        txns.map(async (txn, i) => {
          if (indexesToSign && !indexesToSign.includes(i)) return null;
          const decoded = decodeTransaction2(txn);
          const msg = bytesForSigning.transaction(decoded);
          const sig = await rawEd25519Signer(msg);
          return encodeSignedTransaction({ txn: decoded, sig });
        })
      );
    }
  };
}
function isTestnet(network) {
  return network === ALGORAND_TESTNET_CAIP2 || network === V1_ALGORAND_TESTNET;
}
function toFacilitatorAvmSigner(privateKeyBase64, config) {
  const { address, rawEd25519Signer } = decodePrivateKey(privateKeyBase64);
  const getAlgorandClientForNetwork = (network) => {
    if (isTestnet(network)) {
      if (config?.testnetUrl) {
        return AlgorandClient.fromConfig({
          algodConfig: { server: config.testnetUrl, token: config.algodToken ?? "" }
        });
      }
      return AlgorandClient.testNet();
    }
    if (config?.mainnetUrl) {
      return AlgorandClient.fromConfig({
        algodConfig: { server: config.mainnetUrl, token: config.algodToken ?? "" }
      });
    }
    return AlgorandClient.mainNet();
  };
  const clientCache = /* @__PURE__ */ new Map();
  const getClient = (network) => {
    const key = isTestnet(network) ? "testnet" : "mainnet";
    let client = clientCache.get(key);
    if (!client) {
      client = getAlgorandClientForNetwork(network);
      clientCache.set(key, client);
    }
    return client;
  };
  return {
    getAddresses: () => [address],
    signTransaction: async (txn, _) => {
      const decoded = decodeTransaction2(txn);
      const msg = bytesForSigning.transaction(decoded);
      const sig = await rawEd25519Signer(msg);
      return encodeSignedTransaction({ txn: decoded, sig });
    },
    getAlgodClient: (network) => getClient(network).client.algod,
    simulateTransactions: async (txns, network) => {
      const algod = getClient(network).client.algod;
      return await algod.simulateRawTransactions(txns);
    },
    sendTransactions: async (signedTxns, network) => {
      const algod = getClient(network).client.algod;
      const response = await algod.sendRawTransaction(signedTxns);
      return response.txId;
    },
    waitForConfirmation: async (txId, network, waitRounds = 4) => {
      const algod = getClient(network).client.algod;
      return await waitForConfirmation(txId, waitRounds, algod);
    }
  };
}
export {
  ALGORAND_ADDRESS_LENGTH,
  ALGORAND_ADDRESS_REGEX,
  ALGORAND_MAINNET_CAIP2,
  ALGORAND_MAINNET_GENESIS_HASH,
  ALGORAND_TESTNET_CAIP2,
  ALGORAND_TESTNET_GENESIS_HASH,
  CAIP2_NETWORKS,
  CAIP2_TO_V1,
  DEFAULT_ALGOD_MAINNET,
  DEFAULT_ALGOD_TESTNET,
  ExactAvmScheme,
  MAX_ATOMIC_GROUP_SIZE,
  MAX_REASONABLE_FEE,
  MIN_TXN_FEE,
  NETWORK_TO_ALGOD,
  USDC_CONFIG,
  USDC_DECIMALS,
  USDC_MAINNET_ASA_ID,
  USDC_TESTNET_ASA_ID,
  V1_ALGORAND_MAINNET,
  V1_ALGORAND_TESTNET,
  V1_NETWORKS,
  V1_TO_CAIP2,
  assignGroupId,
  caip2ToV1,
  convertFromTokenAmount,
  convertToTokenAmount,
  createAlgodClient,
  decodeSignedTransaction,
  decodeTransaction,
  decodeUnsignedTransaction,
  encodeTransaction,
  getGenesisHashFromTransaction,
  getNetworkFromCaip2,
  getSenderFromTransaction,
  getTransactionId,
  hasSignature,
  isAlgorandNetwork,
  isAvmSignerWallet,
  isExactAvmPayload,
  isTestnetNetwork,
  isValidAlgorandAddress,
  toClientAvmSigner,
  toFacilitatorAvmSigner,
  v1ToCaip2,
  validateGroupId
};
