// @x402-avm/avm/dist/cjs/exact/facilitator/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/exact/facilitator/index.ts
var facilitator_exports = {};
__export(facilitator_exports, {
  ExactAvmScheme: () => ExactAvmScheme,
  VerifyErrorReason: () => VerifyErrorReason,
  registerExactAvmScheme: () => registerExactAvmScheme
});
module.exports = __toCommonJS(facilitator_exports);

// src/exact/facilitator/scheme.ts
var import_transact2 = require("@algorandfoundation/algokit-utils/transact");
var import_crypto = require("@algorandfoundation/algokit-utils/crypto");

// src/types.ts
function isExactAvmPayload(payload) {
  return typeof payload === "object" && payload !== null && "paymentGroup" in payload && "paymentIndex" in payload && Array.isArray(payload.paymentGroup) && typeof payload.paymentIndex === "number" && payload.paymentGroup.every((item) => typeof item === "string");
}

// src/utils.ts
var import_algod_client = require("@algorandfoundation/algokit-utils/algod-client");
var import_transact = require("@algorandfoundation/algokit-utils/transact");
var import_common = require("@algorandfoundation/algokit-utils/common");

// src/constants.ts
var ALGORAND_MAINNET_CAIP2 = "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
var ALGORAND_TESTNET_CAIP2 = "algorand:SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
var V1_ALGORAND_MAINNET = "algorand-mainnet";
var V1_ALGORAND_TESTNET = "algorand-testnet";
var V1_TO_CAIP2 = {
  [V1_ALGORAND_MAINNET]: ALGORAND_MAINNET_CAIP2,
  [V1_ALGORAND_TESTNET]: ALGORAND_TESTNET_CAIP2
};
var CAIP2_TO_V1 = {
  [ALGORAND_MAINNET_CAIP2]: V1_ALGORAND_MAINNET,
  [ALGORAND_TESTNET_CAIP2]: V1_ALGORAND_TESTNET
};
var USDC_MAINNET_ASA_ID = "31566704";
var USDC_TESTNET_ASA_ID = "10458941";
var USDC_DECIMALS = 6;
var USDC_CONFIG = {
  [ALGORAND_MAINNET_CAIP2]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [ALGORAND_TESTNET_CAIP2]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  // V1 network mappings
  [V1_ALGORAND_MAINNET]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [V1_ALGORAND_TESTNET]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  }
};
var FALLBACK_ALGOD_MAINNET = "https://mainnet-api.algonode.cloud";
var FALLBACK_ALGOD_TESTNET = "https://testnet-api.algonode.cloud";
var DEFAULT_ALGOD_MAINNET = typeof process !== "undefined" && process.env?.ALGOD_MAINNET_URL || FALLBACK_ALGOD_MAINNET;
var DEFAULT_ALGOD_TESTNET = typeof process !== "undefined" && process.env?.ALGOD_TESTNET_URL || FALLBACK_ALGOD_TESTNET;
var NETWORK_TO_ALGOD = {
  [ALGORAND_MAINNET_CAIP2]: DEFAULT_ALGOD_MAINNET,
  [ALGORAND_TESTNET_CAIP2]: DEFAULT_ALGOD_TESTNET,
  [V1_ALGORAND_MAINNET]: DEFAULT_ALGOD_MAINNET,
  [V1_ALGORAND_TESTNET]: DEFAULT_ALGOD_TESTNET
};
var MAX_ATOMIC_GROUP_SIZE = 16;
var MAX_REASONABLE_FEE = 16e3;

// src/utils.ts
var import_algod_client2 = require("@algorandfoundation/algokit-utils/algod-client");
var import_common2 = require("@algorandfoundation/algokit-utils/common");
function decodeTransaction(encoded) {
  return new Uint8Array(Buffer.from(encoded, "base64"));
}
function v1ToCaip2(v1Network) {
  return V1_TO_CAIP2[v1Network] ?? v1Network;
}
function hasSignature(signedTxnBytes) {
  const signedTxn = (0, import_transact.decodeSignedTransaction)(signedTxnBytes);
  return signedTxn.sig !== void 0 || signedTxn.lsig !== void 0 || signedTxn.msig !== void 0;
}

// src/exact/facilitator/scheme.ts
var VerifyErrorReason = {
  INVALID_PAYLOAD_FORMAT: "Invalid payload format",
  GROUP_SIZE_EXCEEDED: "Transaction group exceeds maximum size",
  INVALID_PAYMENT_INDEX: "Payment index out of bounds",
  INVALID_TRANSACTION: "Invalid transaction encoding",
  INVALID_GROUP_ID: "Transactions have inconsistent group IDs",
  PAYMENT_NOT_ASSET_TRANSFER: "Payment transaction is not an asset transfer",
  AMOUNT_MISMATCH: "Payment amount does not match requirements",
  RECEIVER_MISMATCH: "Payment receiver does not match payTo address",
  ASSET_MISMATCH: "Payment asset does not match requirements",
  INVALID_FEE_PAYER: "Fee payer transaction has invalid parameters",
  FEE_TOO_HIGH: "Fee payer transaction fee exceeds maximum",
  PAYMENT_NOT_SIGNED: "Payment transaction is not signed",
  INVALID_SIGNATURE: "Payment transaction signature does not match sender",
  SIMULATION_FAILED: "Transaction simulation failed",
  FACILITATOR_TRANSFERRING_FUNDS: "Facilitator signer cannot be the payment sender",
  GENESIS_HASH_MISMATCH: "Transaction genesis hash does not match expected network",
  // Security error reasons
  UNSIGNED_NON_FACILITATOR_TXN: "Unsigned transaction from non-facilitator address",
  SECURITY_REKEY_NOT_ALLOWED: "Rekey transactions are not allowed",
  SECURITY_CLOSE_TO_NOT_ALLOWED: "Close-to transactions are not allowed",
  SECURITY_KEYREG_NOT_ALLOWED: "Key registration transactions are not allowed"
};
var ExactAvmScheme = class {
  /**
   * Creates a new ExactAvmScheme facilitator instance.
   *
   * @param signer - The AVM signer for facilitator operations
   */
  constructor(signer) {
    this.signer = signer;
    this.scheme = "exact";
    this.caipFamily = "algorand:*";
  }
  /**
   * Get mechanism-specific extra data for the supported kinds endpoint.
   * For AVM, returns the feePayer address for gasless transactions.
   *
   * @param _ - The network identifier (unused, feePayer is network-agnostic)
   * @returns Extra data with feePayer address
   */
  getExtra(_) {
    const addresses = this.signer.getAddresses();
    if (addresses.length === 0) {
      return void 0;
    }
    const randomIndex = Math.floor(Math.random() * addresses.length);
    return { feePayer: addresses[randomIndex] };
  }
  /**
   * Get signer addresses used by this facilitator.
   * Returns all addresses this facilitator can use for signing fee payer transactions.
   *
   * @param _ - The network identifier (unused, addresses are network-agnostic)
   * @returns Array of facilitator wallet addresses
   */
  getSigners(_) {
    return [...this.signer.getAddresses()];
  }
  /**
   * Verifies a payment payload.
   *
   * Verification steps:
   * 1. Validate payload format and structure
   * 2. Decode and validate transaction group
   * 3. Apply security constraints
   * 4. Verify payment transaction (amount, receiver, asset)
   * 5. Prepare signed group (verify + sign fee payers)
   * 6. Simulate transaction group
   *
   * @param payload - The payment payload to verify
   * @param requirements - The payment requirements
   * @returns Promise resolving to verification response
   */
  async verify(payload, requirements) {
    const rawPayload = payload.payload;
    if (!isExactAvmPayload(rawPayload)) {
      return { isValid: false, invalidReason: VerifyErrorReason.INVALID_PAYLOAD_FORMAT };
    }
    const { paymentGroup, paymentIndex } = rawPayload;
    if (paymentGroup.length > MAX_ATOMIC_GROUP_SIZE) {
      return { isValid: false, invalidReason: VerifyErrorReason.GROUP_SIZE_EXCEEDED };
    }
    if (paymentIndex < 0 || paymentIndex >= paymentGroup.length) {
      return { isValid: false, invalidReason: VerifyErrorReason.INVALID_PAYMENT_INDEX };
    }
    const facilitatorAddresses = this.signer.getAddresses();
    const decoded = this.decodeTransactionGroup(paymentGroup, facilitatorAddresses);
    if ("error" in decoded) return decoded.error;
    const paymentTxn = decoded.txns[paymentIndex].txn;
    const payer = paymentTxn.sender.toString();
    const networkStr = String(requirements.network);
    if (networkStr.startsWith("algorand:")) {
      const expectedGenesisHash = networkStr.slice("algorand:".length);
      for (const stxn of decoded.txns) {
        const txnGenesisHash = stxn.txn.genesisHash ? Buffer.from(stxn.txn.genesisHash).toString("base64") : "";
        if (txnGenesisHash !== expectedGenesisHash) {
          return { isValid: false, invalidReason: VerifyErrorReason.GENESIS_HASH_MISMATCH };
        }
      }
    }
    if (facilitatorAddresses.includes(payer)) {
      return { isValid: false, invalidReason: VerifyErrorReason.FACILITATOR_TRANSFERRING_FUNDS };
    }
    const securityCheck = this.verifySecurityConstraints(decoded.txns);
    if (!securityCheck.isValid) return securityCheck;
    const paymentCheck = await this.verifyPaymentTransaction(
      decoded.txns[paymentIndex],
      requirements,
      paymentGroup[paymentIndex]
    );
    if (!paymentCheck.isValid) return paymentCheck;
    const prepared = await this.prepareSignedGroup(decoded.txns, paymentGroup);
    if ("error" in prepared) return prepared.error;
    const simResult = await this.simulateTransactionGroup(prepared.signedTxns, requirements.network);
    if (!simResult.isValid) return simResult;
    return { isValid: true, payer };
  }
  /**
   * Settles a payment by submitting the transaction group.
   *
   * Settlement steps:
   * 1. Verify the payment first
   * 2. Sign fee payer transactions
   * 3. Submit transaction group
   * 4. Return transaction ID
   *
   * Note: Algorand has instant finality, so no confirmation wait is needed.
   *
   * @param payload - The payment payload to settle
   * @param requirements - The payment requirements
   * @returns Promise resolving to settlement response
   */
  async settle(payload, requirements) {
    const verification = await this.verify(payload, requirements);
    if (!verification.isValid) {
      return {
        success: false,
        errorReason: verification.invalidReason,
        transaction: "",
        network: requirements.network,
        payer: verification.payer
      };
    }
    const avmPayload = payload.payload;
    const { paymentGroup, paymentIndex } = avmPayload;
    const facilitatorAddresses = this.signer.getAddresses();
    const signedTxns = [];
    for (let i = 0; i < paymentGroup.length; i++) {
      const txnBytes = decodeTransaction(paymentGroup[i]);
      let txn;
      let isAlreadySigned = false;
      try {
        const stxn = (0, import_transact2.decodeSignedTransaction)(txnBytes);
        if (!stxn.txn.type || stxn.txn.type === "unknown") {
          throw new Error("Invalid signed transaction: missing type");
        }
        txn = stxn.txn;
        isAlreadySigned = stxn.sig !== void 0 || stxn.lsig !== void 0 || stxn.msig !== void 0;
      } catch {
        txn = (0, import_transact2.decodeTransaction)(txnBytes);
        isAlreadySigned = false;
      }
      const sender = txn.sender.toString();
      if (facilitatorAddresses.includes(sender)) {
        const unsignedTxn = (0, import_transact2.encodeTransactionRaw)(txn);
        const signedTxn = await this.signer.signTransaction(unsignedTxn, sender);
        signedTxns.push(signedTxn);
      } else if (isAlreadySigned) {
        signedTxns.push(txnBytes);
      } else {
        return {
          success: false,
          errorReason: `Transaction at index ${i} is unsigned but sender ${sender} is not a facilitator address`,
          transaction: "",
          network: requirements.network
        };
      }
    }
    try {
      await this.signer.sendTransactions(signedTxns, requirements.network);
      const paymentTxnBytes = signedTxns[paymentIndex];
      const paymentStxn = (0, import_transact2.decodeSignedTransaction)(paymentTxnBytes);
      const paymentTxId = paymentStxn.txn.txId();
      return {
        success: true,
        transaction: paymentTxId,
        network: requirements.network,
        payer: verification.payer
      };
    } catch (error) {
      return {
        success: false,
        errorReason: `Failed to submit transaction: ${error instanceof Error ? error.message : "Unknown error"}`,
        transaction: "",
        network: requirements.network,
        payer: verification.payer
      };
    }
  }
  /**
   * Decodes all transactions in the group and validates structure.
   *
   * - Signed transactions are decoded as-is
   * - Unsigned transactions are only accepted from facilitator addresses (fee payers)
   * - Verifies group ID consistency across all transactions
   *
   * @param paymentGroup - Array of base64-encoded transaction strings
   * @param facilitatorAddresses - Addresses controlled by this facilitator
   * @returns Decoded transactions or an error response
   */
  decodeTransactionGroup(paymentGroup, facilitatorAddresses) {
    const txns = [];
    for (let i = 0; i < paymentGroup.length; i++) {
      try {
        const bytes = decodeTransaction(paymentGroup[i]);
        try {
          const stxn = (0, import_transact2.decodeSignedTransaction)(bytes);
          if (!stxn.txn.type || stxn.txn.type === "unknown") {
            throw new Error("Invalid signed transaction: missing type");
          }
          txns.push(stxn);
        } catch {
          const unsignedTxn = (0, import_transact2.decodeTransaction)(bytes);
          const sender = unsignedTxn.sender.toString();
          if (!facilitatorAddresses.includes(sender)) {
            return {
              error: {
                isValid: false,
                invalidReason: `${VerifyErrorReason.UNSIGNED_NON_FACILITATOR_TXN}: transaction at index ${i} from ${sender}`
              }
            };
          }
          const encodedForSimulate = (0, import_transact2.encodeSignedTransaction)({ txn: unsignedTxn });
          txns.push((0, import_transact2.decodeSignedTransaction)(encodedForSimulate));
        }
      } catch {
        return {
          error: {
            isValid: false,
            invalidReason: `${VerifyErrorReason.INVALID_TRANSACTION}: Failed to decode transaction at index ${i}`
          }
        };
      }
    }
    if (txns.length > 1) {
      const firstGroup = txns[0].txn.group;
      const firstGroupId = firstGroup ? Buffer.from(firstGroup).toString("base64") : null;
      for (let i = 1; i < txns.length; i++) {
        const group = txns[i].txn.group;
        const groupId = group ? Buffer.from(group).toString("base64") : null;
        if (groupId !== firstGroupId) {
          return {
            error: { isValid: false, invalidReason: VerifyErrorReason.INVALID_GROUP_ID }
          };
        }
      }
    }
    return { txns };
  }
  /**
   * Verifies fee payer transactions and signs them, returning the assembled group
   * ready for simulation or submission.
   *
   * @param decodedTxns - Decoded signed transaction objects
   * @param paymentGroup - Original base64-encoded transaction strings
   * @returns Signed transaction bytes or an error response
   */
  async prepareSignedGroup(decodedTxns, paymentGroup) {
    const facilitatorAddresses = this.signer.getAddresses();
    const signedTxns = [];
    for (let i = 0; i < decodedTxns.length; i++) {
      const txn = decodedTxns[i].txn;
      const sender = txn.sender.toString();
      if (facilitatorAddresses.includes(sender)) {
        const feeCheck = this.verifyFeePayerTransaction(txn);
        if (!feeCheck.isValid) return { error: feeCheck };
        try {
          const signedTxn = await this.signer.signTransaction((0, import_transact2.encodeTransactionRaw)(txn), sender);
          signedTxns.push(signedTxn);
        } catch (error) {
          return {
            error: {
              isValid: false,
              invalidReason: `Failed to sign fee payer transaction: ${error instanceof Error ? error.message : "Unknown error"}`
            }
          };
        }
      } else {
        signedTxns.push(decodeTransaction(paymentGroup[i]));
      }
    }
    return { signedTxns };
  }
  /**
   * Simulates the transaction group and returns the verification result.
   *
   * @param signedTxns - Signed transaction bytes to simulate
   * @param network - Target network for simulation
   * @returns Verification result from simulation
   */
  async simulateTransactionGroup(signedTxns, network) {
    try {
      const simResult = await this.signer.simulateTransactions(signedTxns, network);
      if (simResult.txnGroups?.[0]?.failureMessage) {
        return {
          isValid: false,
          invalidReason: `${VerifyErrorReason.SIMULATION_FAILED}: ${simResult.txnGroups[0].failureMessage}`
        };
      }
      return { isValid: true };
    } catch (error) {
      return {
        isValid: false,
        invalidReason: `${VerifyErrorReason.SIMULATION_FAILED}: ${error instanceof Error ? error.message : "Unknown error"}`
      };
    }
  }
  /**
   * Verifies the payment transaction matches requirements
   *
   * @param stxn - The signed payment transaction
   * @param requirements - Payment requirements to verify against
   * @param encodedTxn - Base64-encoded transaction for signature check
   * @returns Verification result
   */
  async verifyPaymentTransaction(stxn, requirements, encodedTxn) {
    const txn = stxn.txn;
    if (txn.type !== "axfer") {
      return {
        isValid: false,
        invalidReason: VerifyErrorReason.PAYMENT_NOT_ASSET_TRANSFER
      };
    }
    const assetTransfer = txn.assetTransfer;
    if (!assetTransfer) {
      return {
        isValid: false,
        invalidReason: `${VerifyErrorReason.PAYMENT_NOT_ASSET_TRANSFER}: missing assetTransfer data`
      };
    }
    const amount = assetTransfer.amount ?? BigInt(0);
    if (amount !== BigInt(requirements.amount)) {
      return {
        isValid: false,
        invalidReason: `${VerifyErrorReason.AMOUNT_MISMATCH}: expected ${requirements.amount}, got ${amount.toString()}`
      };
    }
    const receiver = assetTransfer.receiver ? assetTransfer.receiver.toString() : "";
    if (receiver !== requirements.payTo) {
      return {
        isValid: false,
        invalidReason: `${VerifyErrorReason.RECEIVER_MISMATCH}: expected ${requirements.payTo}, got ${receiver}`
      };
    }
    const assetId = assetTransfer.assetId?.toString() ?? "";
    if (assetId !== requirements.asset) {
      return {
        isValid: false,
        invalidReason: `${VerifyErrorReason.ASSET_MISMATCH}: expected ${requirements.asset}, got ${assetId}`
      };
    }
    const txnBytes = decodeTransaction(encodedTxn);
    if (!hasSignature(txnBytes)) {
      return {
        isValid: false,
        invalidReason: VerifyErrorReason.PAYMENT_NOT_SIGNED
      };
    }
    if (stxn.sig) {
      const signedMsg = import_transact2.bytesForSigning.transaction(txn);
      const isValidSig = await (0, import_crypto.ed25519Verifier)(stxn.sig, signedMsg, txn.sender.publicKey);
      if (!isValidSig) {
        return {
          isValid: false,
          invalidReason: VerifyErrorReason.INVALID_SIGNATURE
        };
      }
    }
    return { isValid: true };
  }
  /**
   * Verifies a fee payer transaction is safe to sign
   *
   * @param txn - The fee payer transaction to validate
   * @returns Verification result
   */
  verifyFeePayerTransaction(txn) {
    if (txn.type !== "pay") {
      return {
        isValid: false,
        invalidReason: `${VerifyErrorReason.INVALID_FEE_PAYER}: expected payment transaction, got ${txn.type}`
      };
    }
    const paymentFields = txn.payment;
    const payAmount = paymentFields?.amount ?? BigInt(0);
    if (payAmount > BigInt(0)) {
      return {
        isValid: false,
        invalidReason: `${VerifyErrorReason.INVALID_FEE_PAYER}: amount must be 0`
      };
    }
    if (paymentFields?.receiver) {
      const receiverAddr = paymentFields.receiver.toString();
      const senderAddr = txn.sender.toString();
      if (receiverAddr !== senderAddr) {
        return {
          isValid: false,
          invalidReason: `${VerifyErrorReason.INVALID_FEE_PAYER}: receiver must be same as sender (self-payment)`
        };
      }
    }
    if (paymentFields?.closeRemainderTo) {
      return {
        isValid: false,
        invalidReason: `${VerifyErrorReason.INVALID_FEE_PAYER}: closeRemainderTo not allowed`
      };
    }
    if (txn.rekeyTo) {
      return {
        isValid: false,
        invalidReason: `${VerifyErrorReason.INVALID_FEE_PAYER}: rekeyTo not allowed`
      };
    }
    const fee = Number(txn.fee ?? 0);
    if (fee > MAX_REASONABLE_FEE) {
      return {
        isValid: false,
        invalidReason: `${VerifyErrorReason.FEE_TOO_HIGH}: ${fee} exceeds maximum ${MAX_REASONABLE_FEE}`
      };
    }
    return { isValid: true };
  }
  /**
   * Verifies security constraints for ALL transactions in the group.
   *
   * Security checks:
   * 1. No keyreg (key registration) transactions allowed
   * 2. No rekey transactions allowed
   * 3. No close-to or close-remainder-to fields allowed on any transaction
   *
   * @param txns - Signed transactions to check for security violations
   * @returns Verification result
   */
  verifySecurityConstraints(txns) {
    for (let i = 0; i < txns.length; i++) {
      const txn = txns[i].txn;
      if (txn.type === "keyreg") {
        return {
          isValid: false,
          invalidReason: `${VerifyErrorReason.SECURITY_KEYREG_NOT_ALLOWED}: Transaction at index ${i} is a key registration transaction`
        };
      }
      if (txn.rekeyTo) {
        return {
          isValid: false,
          invalidReason: `${VerifyErrorReason.SECURITY_REKEY_NOT_ALLOWED}: Transaction at index ${i} has rekeyTo set`
        };
      }
      if (txn.type === "pay") {
        if (txn.payment?.closeRemainderTo) {
          return {
            isValid: false,
            invalidReason: `${VerifyErrorReason.SECURITY_CLOSE_TO_NOT_ALLOWED}: Transaction at index ${i} has CloseRemainderTo set`
          };
        }
      }
      if (txn.type === "axfer") {
        if (txn.assetTransfer?.closeRemainderTo) {
          return {
            isValid: false,
            invalidReason: `${VerifyErrorReason.SECURITY_CLOSE_TO_NOT_ALLOWED}: Transaction at index ${i} has AssetCloseTo set`
          };
        }
      }
    }
    return { isValid: true };
  }
};

// src/exact/v1/facilitator/scheme.ts
var import_transact3 = require("@algorandfoundation/algokit-utils/transact");
var ExactAvmSchemeV1 = class {
  /**
   * Creates a new ExactAvmSchemeV1 facilitator instance.
   *
   * @param signer - The AVM signer for facilitator operations
   */
  constructor(signer) {
    this.signer = signer;
    this.scheme = "exact";
    this.caipFamily = "algorand:*";
  }
  /**
   * Get mechanism-specific extra data for the supported kinds endpoint.
   *
   * @param _ - The network identifier (unused)
   * @returns Extra data with feePayer address
   */
  getExtra(_) {
    const addresses = this.signer.getAddresses();
    if (addresses.length === 0) {
      return void 0;
    }
    const randomIndex = Math.floor(Math.random() * addresses.length);
    return { feePayer: addresses[randomIndex] };
  }
  /**
   * Get signer addresses used by this facilitator.
   *
   * @param _ - The network identifier (unused)
   * @returns Array of facilitator wallet addresses
   */
  getSigners(_) {
    return [...this.signer.getAddresses()];
  }
  /**
   * Verifies a V1 payment payload.
   *
   * @param payload - The payment payload to verify
   * @param requirements - The payment requirements
   * @returns Promise resolving to verification response
   */
  async verify(payload, requirements) {
    const rawPayload = payload.payload;
    if (!isExactAvmPayload(rawPayload)) {
      return {
        isValid: false,
        invalidReason: "Invalid payload format"
      };
    }
    const avmPayload = rawPayload;
    const { paymentGroup, paymentIndex } = avmPayload;
    if (paymentGroup.length > MAX_ATOMIC_GROUP_SIZE) {
      return {
        isValid: false,
        invalidReason: "Transaction group exceeds maximum size"
      };
    }
    if (paymentIndex < 0 || paymentIndex >= paymentGroup.length) {
      return {
        isValid: false,
        invalidReason: "Payment index out of bounds"
      };
    }
    let paymentTxn;
    try {
      const txnBytes2 = decodeTransaction(paymentGroup[paymentIndex]);
      paymentTxn = (0, import_transact3.decodeSignedTransaction)(txnBytes2);
    } catch (error) {
      return {
        isValid: false,
        invalidReason: `Invalid transaction encoding: ${error instanceof Error ? error.message : "Unknown error"}`
      };
    }
    const txn = paymentTxn.txn;
    if (txn.type !== "axfer") {
      return {
        isValid: false,
        invalidReason: "Payment transaction is not an asset transfer"
      };
    }
    const assetTransfer = txn.assetTransfer;
    const amount = (assetTransfer?.amount ?? BigInt(0)).toString();
    const requiredAmount = requirements.maxAmountRequired ?? requirements.amount;
    if (amount !== requiredAmount) {
      return {
        isValid: false,
        invalidReason: `Amount mismatch: expected ${requiredAmount}, got ${amount}`
      };
    }
    const receiver = assetTransfer?.receiver ? assetTransfer.receiver.toString() : "";
    if (receiver !== requirements.payTo) {
      return {
        isValid: false,
        invalidReason: `Receiver mismatch: expected ${requirements.payTo}, got ${receiver}`
      };
    }
    const assetId = assetTransfer?.assetId?.toString() ?? "";
    if (assetId !== requirements.asset) {
      return {
        isValid: false,
        invalidReason: `Asset mismatch: expected ${requirements.asset}, got ${assetId}`
      };
    }
    const txnBytes = decodeTransaction(paymentGroup[paymentIndex]);
    if (!hasSignature(txnBytes)) {
      return {
        isValid: false,
        invalidReason: "Payment transaction is not signed"
      };
    }
    return { isValid: true };
  }
  /**
   * Settles a V1 payment.
   *
   * @param payload - The payment payload to settle
   * @param requirements - The payment requirements
   * @returns Promise resolving to settlement response
   */
  async settle(payload, requirements) {
    const verification = await this.verify(payload, requirements);
    if (!verification.isValid) {
      return {
        success: false,
        errorReason: verification.invalidReason,
        transaction: "",
        network: requirements.network
      };
    }
    const avmPayload = payload.payload;
    const { paymentGroup, paymentIndex } = avmPayload;
    const caip2Network = v1ToCaip2(requirements.network);
    const signedTxns = paymentGroup.map((encoded) => decodeTransaction(encoded));
    try {
      await this.signer.sendTransactions(signedTxns, caip2Network);
      const paymentStxn = (0, import_transact3.decodeSignedTransaction)(signedTxns[paymentIndex]);
      const paymentTxId = paymentStxn.txn.txId();
      return {
        success: true,
        transaction: paymentTxId,
        network: requirements.network
      };
    } catch (error) {
      return {
        success: false,
        errorReason: `Failed to submit transaction: ${error instanceof Error ? error.message : "Unknown error"}`,
        transaction: "",
        network: requirements.network
      };
    }
  }
};

// src/exact/v1/client/scheme.ts
var import_transact4 = require("@algorandfoundation/algokit-utils/transact");
var import_common3 = require("@algorandfoundation/algokit-utils/common");

// src/v1/index.ts
var AVM_NETWORK_CAIP2_MAP = {
  "algorand-mainnet": "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=",
  "algorand-testnet": "algorand:SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI="
};
var NETWORKS = Object.keys(AVM_NETWORK_CAIP2_MAP);

// src/exact/facilitator/register.ts
function registerExactAvmScheme(facilitator, config) {
  facilitator.register(config.networks, new ExactAvmScheme(config.signer));
  facilitator.registerV1(NETWORKS, new ExactAvmSchemeV1(config.signer));
  return facilitator;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ExactAvmScheme,
  VerifyErrorReason,
  registerExactAvmScheme
});
