// @x402-avm/avm/dist/esm/chunk-TLHHCNYP.mjs
import {
  ALGORAND_ADDRESS_REGEX,
  ALGORAND_MAINNET_GENESIS_HASH,
  ALGORAND_TESTNET_CAIP2,
  ALGORAND_TESTNET_GENESIS_HASH,
  CAIP2_TO_V1,
  DEFAULT_ALGOD_TESTNET,
  NETWORK_TO_ALGOD,
  V1_ALGORAND_MAINNET,
  V1_ALGORAND_TESTNET,
  V1_TO_CAIP2
} from "./chunk-DUGAE7GK.mjs";

// src/utils.ts
import { AlgodClient } from "@algorandfoundation/algokit-utils/algod-client";
import {
  decodeTransaction as decodeUnsignedTxn,
  decodeSignedTransaction as decodeSignedTxn,
  groupTransactions
} from "@algorandfoundation/algokit-utils/transact";
import { isValidAddress } from "@algorandfoundation/algokit-utils/common";
import { AlgodClient as AlgodClient2 } from "@algorandfoundation/algokit-utils/algod-client";
import { Address, encodeAddress, decodeAddress } from "@algorandfoundation/algokit-utils/common";
function createAlgodClient(network, customUrl, token = "") {
  const url = customUrl ?? NETWORK_TO_ALGOD[network] ?? DEFAULT_ALGOD_TESTNET;
  return new AlgodClient({ baseUrl: url, token: token || void 0 });
}
function encodeTransaction(txn) {
  return Buffer.from(txn).toString("base64");
}
function decodeTransaction(encoded) {
  return new Uint8Array(Buffer.from(encoded, "base64"));
}
function decodeSignedTransaction(encoded) {
  const bytes = decodeTransaction(encoded);
  return decodeSignedTxn(bytes);
}
function decodeUnsignedTransaction(encoded) {
  const bytes = decodeTransaction(encoded);
  return decodeUnsignedTxn(bytes);
}
function isValidAlgorandAddress(address) {
  if (!ALGORAND_ADDRESS_REGEX.test(address)) {
    return false;
  }
  return isValidAddress(address);
}
function getSenderFromTransaction(txnBytes, isSigned = true) {
  if (isSigned) {
    const signedTxn = decodeSignedTxn(txnBytes);
    return signedTxn.txn.sender.toString();
  }
  const txn = decodeUnsignedTxn(txnBytes);
  return txn.sender.toString();
}
function convertToTokenAmount(decimalAmount, decimals) {
  const amount = parseFloat(decimalAmount);
  if (isNaN(amount)) {
    throw new Error(`Invalid amount: ${decimalAmount}`);
  }
  const [intPart, decPart = ""] = String(amount).split(".");
  const paddedDec = decPart.padEnd(decimals, "0").slice(0, decimals);
  const tokenAmount = (intPart + paddedDec).replace(/^0+/, "") || "0";
  return tokenAmount;
}
function convertFromTokenAmount(atomicAmount, decimals) {
  const amount = BigInt(atomicAmount);
  const divisor = BigInt(10 ** decimals);
  const intPart = amount / divisor;
  const decPart = amount % divisor;
  if (decPart === BigInt(0)) {
    return intPart.toString();
  }
  const decStr = decPart.toString().padStart(decimals, "0");
  const trimmedDec = decStr.replace(/0+$/, "");
  return `${intPart}.${trimmedDec}`;
}
function getNetworkFromCaip2(caip2) {
  if (!caip2.startsWith("algorand:")) {
    return null;
  }
  const genesisHash = caip2.slice("algorand:".length);
  if (genesisHash === ALGORAND_MAINNET_GENESIS_HASH) {
    return "mainnet";
  }
  if (genesisHash === ALGORAND_TESTNET_GENESIS_HASH) {
    return "testnet";
  }
  return null;
}
function isAlgorandNetwork(network) {
  if (network.startsWith("algorand:")) {
    return true;
  }
  return network === V1_ALGORAND_MAINNET || network === V1_ALGORAND_TESTNET;
}
function isTestnetNetwork(network) {
  return network === ALGORAND_TESTNET_CAIP2 || network === V1_ALGORAND_TESTNET;
}
function v1ToCaip2(v1Network) {
  return V1_TO_CAIP2[v1Network] ?? v1Network;
}
function caip2ToV1(caip2Network) {
  return CAIP2_TO_V1[caip2Network] ?? caip2Network;
}
function getGenesisHashFromTransaction(txn) {
  if (!txn.genesisHash) {
    throw new Error("Transaction does not have a genesis hash");
  }
  return Buffer.from(txn.genesisHash).toString("base64");
}
function validateGroupId(txns) {
  if (txns.length <= 1) {
    return true;
  }
  let expectedGroupId = null;
  for (const txnBytes of txns) {
    const txn = decodeUnsignedTxn(txnBytes);
    const groupId = txn.group ? Buffer.from(txn.group).toString("base64") : null;
    if (expectedGroupId === null) {
      expectedGroupId = groupId;
    } else if (groupId !== expectedGroupId) {
      return false;
    }
  }
  return true;
}
function assignGroupId(txns) {
  if (txns.length <= 1) {
    return txns;
  }
  return groupTransactions(txns);
}
function getTransactionId(signedTxnBytes) {
  const signedTxn = decodeSignedTxn(signedTxnBytes);
  return signedTxn.txn.txId();
}
function hasSignature(signedTxnBytes) {
  const signedTxn = decodeSignedTxn(signedTxnBytes);
  return signedTxn.sig !== void 0 || signedTxn.lsig !== void 0 || signedTxn.msig !== void 0;
}

export {
  createAlgodClient,
  encodeTransaction,
  decodeTransaction,
  decodeSignedTransaction,
  decodeUnsignedTransaction,
  isValidAlgorandAddress,
  getSenderFromTransaction,
  convertToTokenAmount,
  convertFromTokenAmount,
  getNetworkFromCaip2,
  isAlgorandNetwork,
  isTestnetNetwork,
  v1ToCaip2,
  caip2ToV1,
  getGenesisHashFromTransaction,
  validateGroupId,
  assignGroupId,
  getTransactionId,
  hasSignature
};
