// @x402-avm/avm/dist/esm/chunk-ZXLI3SMZ.mjs
import {
  createAlgodClient,
  encodeTransaction
} from "./chunk-TLHHCNYP.mjs";
import {
  DEFAULT_ALGOD_TESTNET,
  USDC_CONFIG
} from "./chunk-DUGAE7GK.mjs";

// src/exact/client/scheme.ts
import {
  Transaction,
  TransactionType,
  encodeTransactionRaw,
  groupTransactions
} from "@algorandfoundation/algokit-utils/transact";
import { Address } from "@algorandfoundation/algokit-utils/common";
var ExactAvmScheme = class {
  /**
   * Creates a new ExactAvmScheme instance.
   *
   * @param signer - The AVM signer for client operations
   * @param config - Optional configuration for Algod client
   */
  constructor(signer, config) {
    this.signer = signer;
    this.config = config;
    this.scheme = "exact";
  }
  /**
   * Creates a payment payload for the Exact scheme.
   *
   * Constructs an atomic transaction group with:
   * - Optional fee payer transaction (if feePayer specified in requirements.extra)
   * - ASA transfer transaction to payTo address
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements
   * @returns Promise resolving to a payment payload result
   */
  async createPaymentPayload(x402Version, paymentRequirements) {
    const { amount, asset, payTo, network, extra } = paymentRequirements;
    const algodClient = this.config?.algodClient ?? createAlgodClient(
      network,
      this.config?.algodUrl ?? DEFAULT_ALGOD_TESTNET,
      this.config?.algodToken
    );
    const suggestedParams = await algodClient.suggestedParams();
    const assetId = this.getAssetId(asset, network);
    const feePayer = extra?.feePayer;
    const transactions = [];
    let paymentIndex = 0;
    const totalTxnCount = feePayer ? 2 : 1;
    const minFee = suggestedParams.minFee ?? BigInt(1e3);
    if (feePayer) {
      const feePayerTxn = new Transaction({
        type: TransactionType.Payment,
        sender: Address.fromString(feePayer),
        fee: minFee * BigInt(totalTxnCount),
        firstValid: suggestedParams.firstValid,
        lastValid: suggestedParams.lastValid,
        genesisHash: suggestedParams.genesisHash,
        genesisId: suggestedParams.genesisId,
        note: new Uint8Array(Buffer.from(`x402-fee-payer-${Date.now()}`)),
        payment: {
          receiver: Address.fromString(feePayer),
          amount: BigInt(0)
        }
      });
      transactions.push(feePayerTxn);
      paymentIndex = 1;
    }
    const assetTransferFee = feePayer ? BigInt(0) : suggestedParams.fee ?? minFee;
    const assetTransferTxn = new Transaction({
      type: TransactionType.AssetTransfer,
      sender: Address.fromString(this.signer.address),
      fee: assetTransferFee,
      firstValid: suggestedParams.firstValid,
      lastValid: suggestedParams.lastValid,
      genesisHash: suggestedParams.genesisHash,
      genesisId: suggestedParams.genesisId,
      note: new Uint8Array(Buffer.from(`x402-payment-v${x402Version}-${Date.now()}`)),
      assetTransfer: {
        receiver: Address.fromString(payTo),
        amount: BigInt(amount),
        assetId: BigInt(assetId)
      }
    });
    transactions.push(assetTransferTxn);
    let groupedTxns = transactions;
    if (transactions.length > 1) {
      groupedTxns = groupTransactions(transactions);
    }
    const encodedTxns = groupedTxns.map((txn) => encodeTransactionRaw(txn));
    const clientIndexes = groupedTxns.map((txn, i) => {
      const sender = txn.sender.toString();
      return sender === this.signer.address ? i : -1;
    }).filter((i) => i !== -1);
    console.log("[x402 AVM Client] Creating payment:", {
      sender: this.signer.address,
      receiver: payTo,
      amount,
      assetId: this.getAssetId(asset, network),
      network,
      clientIndexes,
      txnCount: groupedTxns.length,
      hasFeePayer: !!feePayer
    });
    const signedTxns = await this.signer.signTransactions(encodedTxns, clientIndexes);
    console.log("[x402 AVM Client] Signed transactions:", {
      signedCount: signedTxns.filter((t) => t !== null).length,
      totalCount: signedTxns.length,
      signedIndexes: signedTxns.map((t, i) => t !== null ? i : -1).filter((i) => i !== -1)
    });
    const paymentGroup = encodedTxns.map((txnBytes, i) => {
      const signedTxn = signedTxns[i];
      if (signedTxn) {
        return encodeTransaction(signedTxn);
      }
      return encodeTransaction(txnBytes);
    });
    const payload = {
      paymentGroup,
      paymentIndex
    };
    return {
      x402Version,
      payload
    };
  }
  /**
   * Gets the asset ID from the requirements or defaults to USDC
   *
   * @param asset - Asset identifier from requirements
   * @param network - Network identifier
   * @returns Asset ID as string
   */
  getAssetId(asset, network) {
    if (/^\d+$/.test(asset)) {
      return asset;
    }
    const usdcConfig = USDC_CONFIG[network];
    if (usdcConfig) {
      return usdcConfig.asaId;
    }
    return asset;
  }
};

export {
  ExactAvmScheme
};
