// @x402-avm/avm/dist/esm/chunk-DUGAE7GK.mjs
// src/constants.ts
var ALGORAND_MAINNET_CAIP2 = "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
var ALGORAND_TESTNET_CAIP2 = "algorand:SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
var CAIP2_NETWORKS = [ALGORAND_MAINNET_CAIP2, ALGORAND_TESTNET_CAIP2];
var ALGORAND_MAINNET_GENESIS_HASH = "wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
var ALGORAND_TESTNET_GENESIS_HASH = "SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
var V1_ALGORAND_MAINNET = "algorand-mainnet";
var V1_ALGORAND_TESTNET = "algorand-testnet";
var V1_NETWORKS = [V1_ALGORAND_MAINNET, V1_ALGORAND_TESTNET];
var V1_TO_CAIP2 = {
  [V1_ALGORAND_MAINNET]: ALGORAND_MAINNET_CAIP2,
  [V1_ALGORAND_TESTNET]: ALGORAND_TESTNET_CAIP2
};
var CAIP2_TO_V1 = {
  [ALGORAND_MAINNET_CAIP2]: V1_ALGORAND_MAINNET,
  [ALGORAND_TESTNET_CAIP2]: V1_ALGORAND_TESTNET
};
var USDC_MAINNET_ASA_ID = "31566704";
var USDC_TESTNET_ASA_ID = "10458941";
var USDC_DECIMALS = 6;
var USDC_CONFIG = {
  [ALGORAND_MAINNET_CAIP2]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [ALGORAND_TESTNET_CAIP2]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  // V1 network mappings
  [V1_ALGORAND_MAINNET]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [V1_ALGORAND_TESTNET]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  }
};
var FALLBACK_ALGOD_MAINNET = "https://mainnet-api.algonode.cloud";
var FALLBACK_ALGOD_TESTNET = "https://testnet-api.algonode.cloud";
var DEFAULT_ALGOD_MAINNET = typeof process !== "undefined" && process.env?.ALGOD_MAINNET_URL || FALLBACK_ALGOD_MAINNET;
var DEFAULT_ALGOD_TESTNET = typeof process !== "undefined" && process.env?.ALGOD_TESTNET_URL || FALLBACK_ALGOD_TESTNET;
var NETWORK_TO_ALGOD = {
  [ALGORAND_MAINNET_CAIP2]: DEFAULT_ALGOD_MAINNET,
  [ALGORAND_TESTNET_CAIP2]: DEFAULT_ALGOD_TESTNET,
  [V1_ALGORAND_MAINNET]: DEFAULT_ALGOD_MAINNET,
  [V1_ALGORAND_TESTNET]: DEFAULT_ALGOD_TESTNET
};
var MAX_ATOMIC_GROUP_SIZE = 16;
var MIN_TXN_FEE = 1e3;
var MAX_REASONABLE_FEE = 16e3;
var ALGORAND_ADDRESS_REGEX = /^[A-Z2-7]{58}$/;
var ALGORAND_ADDRESS_LENGTH = 58;

export {
  ALGORAND_MAINNET_CAIP2,
  ALGORAND_TESTNET_CAIP2,
  CAIP2_NETWORKS,
  ALGORAND_MAINNET_GENESIS_HASH,
  ALGORAND_TESTNET_GENESIS_HASH,
  V1_ALGORAND_MAINNET,
  V1_ALGORAND_TESTNET,
  V1_NETWORKS,
  V1_TO_CAIP2,
  CAIP2_TO_V1,
  USDC_MAINNET_ASA_ID,
  USDC_TESTNET_ASA_ID,
  USDC_DECIMALS,
  USDC_CONFIG,
  DEFAULT_ALGOD_MAINNET,
  DEFAULT_ALGOD_TESTNET,
  NETWORK_TO_ALGOD,
  MAX_ATOMIC_GROUP_SIZE,
  MIN_TXN_FEE,
  MAX_REASONABLE_FEE,
  ALGORAND_ADDRESS_REGEX,
  ALGORAND_ADDRESS_LENGTH
};
