// @x402-avm/avm/dist/esm/chunk-EHXVDHNB.mjs
import {
  isExactAvmPayload
} from "./chunk-IHTCCPJQ.mjs";
import {
  decodeTransaction,
  hasSignature,
  v1ToCaip2
} from "./chunk-TLHHCNYP.mjs";
import {
  MAX_ATOMIC_GROUP_SIZE
} from "./chunk-DUGAE7GK.mjs";

// src/exact/v1/facilitator/scheme.ts
import { decodeSignedTransaction as decodeSignedTxn } from "@algorandfoundation/algokit-utils/transact";
var ExactAvmSchemeV1 = class {
  /**
   * Creates a new ExactAvmSchemeV1 facilitator instance.
   *
   * @param signer - The AVM signer for facilitator operations
   */
  constructor(signer) {
    this.signer = signer;
    this.scheme = "exact";
    this.caipFamily = "algorand:*";
  }
  /**
   * Get mechanism-specific extra data for the supported kinds endpoint.
   *
   * @param _ - The network identifier (unused)
   * @returns Extra data with feePayer address
   */
  getExtra(_) {
    const addresses = this.signer.getAddresses();
    if (addresses.length === 0) {
      return void 0;
    }
    const randomIndex = Math.floor(Math.random() * addresses.length);
    return { feePayer: addresses[randomIndex] };
  }
  /**
   * Get signer addresses used by this facilitator.
   *
   * @param _ - The network identifier (unused)
   * @returns Array of facilitator wallet addresses
   */
  getSigners(_) {
    return [...this.signer.getAddresses()];
  }
  /**
   * Verifies a V1 payment payload.
   *
   * @param payload - The payment payload to verify
   * @param requirements - The payment requirements
   * @returns Promise resolving to verification response
   */
  async verify(payload, requirements) {
    const rawPayload = payload.payload;
    if (!isExactAvmPayload(rawPayload)) {
      return {
        isValid: false,
        invalidReason: "Invalid payload format"
      };
    }
    const avmPayload = rawPayload;
    const { paymentGroup, paymentIndex } = avmPayload;
    if (paymentGroup.length > MAX_ATOMIC_GROUP_SIZE) {
      return {
        isValid: false,
        invalidReason: "Transaction group exceeds maximum size"
      };
    }
    if (paymentIndex < 0 || paymentIndex >= paymentGroup.length) {
      return {
        isValid: false,
        invalidReason: "Payment index out of bounds"
      };
    }
    let paymentTxn;
    try {
      const txnBytes2 = decodeTransaction(paymentGroup[paymentIndex]);
      paymentTxn = decodeSignedTxn(txnBytes2);
    } catch (error) {
      return {
        isValid: false,
        invalidReason: `Invalid transaction encoding: ${error instanceof Error ? error.message : "Unknown error"}`
      };
    }
    const txn = paymentTxn.txn;
    if (txn.type !== "axfer") {
      return {
        isValid: false,
        invalidReason: "Payment transaction is not an asset transfer"
      };
    }
    const assetTransfer = txn.assetTransfer;
    const amount = (assetTransfer?.amount ?? BigInt(0)).toString();
    const requiredAmount = requirements.maxAmountRequired ?? requirements.amount;
    if (amount !== requiredAmount) {
      return {
        isValid: false,
        invalidReason: `Amount mismatch: expected ${requiredAmount}, got ${amount}`
      };
    }
    const receiver = assetTransfer?.receiver ? assetTransfer.receiver.toString() : "";
    if (receiver !== requirements.payTo) {
      return {
        isValid: false,
        invalidReason: `Receiver mismatch: expected ${requirements.payTo}, got ${receiver}`
      };
    }
    const assetId = assetTransfer?.assetId?.toString() ?? "";
    if (assetId !== requirements.asset) {
      return {
        isValid: false,
        invalidReason: `Asset mismatch: expected ${requirements.asset}, got ${assetId}`
      };
    }
    const txnBytes = decodeTransaction(paymentGroup[paymentIndex]);
    if (!hasSignature(txnBytes)) {
      return {
        isValid: false,
        invalidReason: "Payment transaction is not signed"
      };
    }
    return { isValid: true };
  }
  /**
   * Settles a V1 payment.
   *
   * @param payload - The payment payload to settle
   * @param requirements - The payment requirements
   * @returns Promise resolving to settlement response
   */
  async settle(payload, requirements) {
    const verification = await this.verify(payload, requirements);
    if (!verification.isValid) {
      return {
        success: false,
        errorReason: verification.invalidReason,
        transaction: "",
        network: requirements.network
      };
    }
    const avmPayload = payload.payload;
    const { paymentGroup, paymentIndex } = avmPayload;
    const caip2Network = v1ToCaip2(requirements.network);
    const signedTxns = paymentGroup.map((encoded) => decodeTransaction(encoded));
    try {
      await this.signer.sendTransactions(signedTxns, caip2Network);
      const paymentStxn = decodeSignedTxn(signedTxns[paymentIndex]);
      const paymentTxId = paymentStxn.txn.txId();
      return {
        success: true,
        transaction: paymentTxId,
        network: requirements.network
      };
    } catch (error) {
      return {
        success: false,
        errorReason: `Failed to submit transaction: ${error instanceof Error ? error.message : "Unknown error"}`,
        transaction: "",
        network: requirements.network
      };
    }
  }
};

export {
  ExactAvmSchemeV1
};
