// @x402-avm/avm/dist/cjs/exact/v1/facilitator/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/exact/v1/facilitator/index.ts
var facilitator_exports = {};
__export(facilitator_exports, {
  ExactAvmSchemeV1: () => ExactAvmSchemeV1
});
module.exports = __toCommonJS(facilitator_exports);

// src/exact/v1/facilitator/scheme.ts
var import_transact2 = require("@algorandfoundation/algokit-utils/transact");

// src/types.ts
function isExactAvmPayload(payload) {
  return typeof payload === "object" && payload !== null && "paymentGroup" in payload && "paymentIndex" in payload && Array.isArray(payload.paymentGroup) && typeof payload.paymentIndex === "number" && payload.paymentGroup.every((item) => typeof item === "string");
}

// src/utils.ts
var import_algod_client = require("@algorandfoundation/algokit-utils/algod-client");
var import_transact = require("@algorandfoundation/algokit-utils/transact");
var import_common = require("@algorandfoundation/algokit-utils/common");

// src/constants.ts
var ALGORAND_MAINNET_CAIP2 = "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
var ALGORAND_TESTNET_CAIP2 = "algorand:SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
var V1_ALGORAND_MAINNET = "algorand-mainnet";
var V1_ALGORAND_TESTNET = "algorand-testnet";
var V1_TO_CAIP2 = {
  [V1_ALGORAND_MAINNET]: ALGORAND_MAINNET_CAIP2,
  [V1_ALGORAND_TESTNET]: ALGORAND_TESTNET_CAIP2
};
var CAIP2_TO_V1 = {
  [ALGORAND_MAINNET_CAIP2]: V1_ALGORAND_MAINNET,
  [ALGORAND_TESTNET_CAIP2]: V1_ALGORAND_TESTNET
};
var USDC_MAINNET_ASA_ID = "31566704";
var USDC_TESTNET_ASA_ID = "10458941";
var USDC_DECIMALS = 6;
var USDC_CONFIG = {
  [ALGORAND_MAINNET_CAIP2]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [ALGORAND_TESTNET_CAIP2]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  // V1 network mappings
  [V1_ALGORAND_MAINNET]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [V1_ALGORAND_TESTNET]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  }
};
var FALLBACK_ALGOD_MAINNET = "https://mainnet-api.algonode.cloud";
var FALLBACK_ALGOD_TESTNET = "https://testnet-api.algonode.cloud";
var DEFAULT_ALGOD_MAINNET = typeof process !== "undefined" && process.env?.ALGOD_MAINNET_URL || FALLBACK_ALGOD_MAINNET;
var DEFAULT_ALGOD_TESTNET = typeof process !== "undefined" && process.env?.ALGOD_TESTNET_URL || FALLBACK_ALGOD_TESTNET;
var NETWORK_TO_ALGOD = {
  [ALGORAND_MAINNET_CAIP2]: DEFAULT_ALGOD_MAINNET,
  [ALGORAND_TESTNET_CAIP2]: DEFAULT_ALGOD_TESTNET,
  [V1_ALGORAND_MAINNET]: DEFAULT_ALGOD_MAINNET,
  [V1_ALGORAND_TESTNET]: DEFAULT_ALGOD_TESTNET
};
var MAX_ATOMIC_GROUP_SIZE = 16;

// src/utils.ts
var import_algod_client2 = require("@algorandfoundation/algokit-utils/algod-client");
var import_common2 = require("@algorandfoundation/algokit-utils/common");
function decodeTransaction(encoded) {
  return new Uint8Array(Buffer.from(encoded, "base64"));
}
function v1ToCaip2(v1Network) {
  return V1_TO_CAIP2[v1Network] ?? v1Network;
}
function hasSignature(signedTxnBytes) {
  const signedTxn = (0, import_transact.decodeSignedTransaction)(signedTxnBytes);
  return signedTxn.sig !== void 0 || signedTxn.lsig !== void 0 || signedTxn.msig !== void 0;
}

// src/exact/v1/facilitator/scheme.ts
var ExactAvmSchemeV1 = class {
  /**
   * Creates a new ExactAvmSchemeV1 facilitator instance.
   *
   * @param signer - The AVM signer for facilitator operations
   */
  constructor(signer) {
    this.signer = signer;
    this.scheme = "exact";
    this.caipFamily = "algorand:*";
  }
  /**
   * Get mechanism-specific extra data for the supported kinds endpoint.
   *
   * @param _ - The network identifier (unused)
   * @returns Extra data with feePayer address
   */
  getExtra(_) {
    const addresses = this.signer.getAddresses();
    if (addresses.length === 0) {
      return void 0;
    }
    const randomIndex = Math.floor(Math.random() * addresses.length);
    return { feePayer: addresses[randomIndex] };
  }
  /**
   * Get signer addresses used by this facilitator.
   *
   * @param _ - The network identifier (unused)
   * @returns Array of facilitator wallet addresses
   */
  getSigners(_) {
    return [...this.signer.getAddresses()];
  }
  /**
   * Verifies a V1 payment payload.
   *
   * @param payload - The payment payload to verify
   * @param requirements - The payment requirements
   * @returns Promise resolving to verification response
   */
  async verify(payload, requirements) {
    const rawPayload = payload.payload;
    if (!isExactAvmPayload(rawPayload)) {
      return {
        isValid: false,
        invalidReason: "Invalid payload format"
      };
    }
    const avmPayload = rawPayload;
    const { paymentGroup, paymentIndex } = avmPayload;
    if (paymentGroup.length > MAX_ATOMIC_GROUP_SIZE) {
      return {
        isValid: false,
        invalidReason: "Transaction group exceeds maximum size"
      };
    }
    if (paymentIndex < 0 || paymentIndex >= paymentGroup.length) {
      return {
        isValid: false,
        invalidReason: "Payment index out of bounds"
      };
    }
    let paymentTxn;
    try {
      const txnBytes2 = decodeTransaction(paymentGroup[paymentIndex]);
      paymentTxn = (0, import_transact2.decodeSignedTransaction)(txnBytes2);
    } catch (error) {
      return {
        isValid: false,
        invalidReason: `Invalid transaction encoding: ${error instanceof Error ? error.message : "Unknown error"}`
      };
    }
    const txn = paymentTxn.txn;
    if (txn.type !== "axfer") {
      return {
        isValid: false,
        invalidReason: "Payment transaction is not an asset transfer"
      };
    }
    const assetTransfer = txn.assetTransfer;
    const amount = (assetTransfer?.amount ?? BigInt(0)).toString();
    const requiredAmount = requirements.maxAmountRequired ?? requirements.amount;
    if (amount !== requiredAmount) {
      return {
        isValid: false,
        invalidReason: `Amount mismatch: expected ${requiredAmount}, got ${amount}`
      };
    }
    const receiver = assetTransfer?.receiver ? assetTransfer.receiver.toString() : "";
    if (receiver !== requirements.payTo) {
      return {
        isValid: false,
        invalidReason: `Receiver mismatch: expected ${requirements.payTo}, got ${receiver}`
      };
    }
    const assetId = assetTransfer?.assetId?.toString() ?? "";
    if (assetId !== requirements.asset) {
      return {
        isValid: false,
        invalidReason: `Asset mismatch: expected ${requirements.asset}, got ${assetId}`
      };
    }
    const txnBytes = decodeTransaction(paymentGroup[paymentIndex]);
    if (!hasSignature(txnBytes)) {
      return {
        isValid: false,
        invalidReason: "Payment transaction is not signed"
      };
    }
    return { isValid: true };
  }
  /**
   * Settles a V1 payment.
   *
   * @param payload - The payment payload to settle
   * @param requirements - The payment requirements
   * @returns Promise resolving to settlement response
   */
  async settle(payload, requirements) {
    const verification = await this.verify(payload, requirements);
    if (!verification.isValid) {
      return {
        success: false,
        errorReason: verification.invalidReason,
        transaction: "",
        network: requirements.network
      };
    }
    const avmPayload = payload.payload;
    const { paymentGroup, paymentIndex } = avmPayload;
    const caip2Network = v1ToCaip2(requirements.network);
    const signedTxns = paymentGroup.map((encoded) => decodeTransaction(encoded));
    try {
      await this.signer.sendTransactions(signedTxns, caip2Network);
      const paymentStxn = (0, import_transact2.decodeSignedTransaction)(signedTxns[paymentIndex]);
      const paymentTxId = paymentStxn.txn.txId();
      return {
        success: true,
        transaction: paymentTxId,
        network: requirements.network
      };
    } catch (error) {
      return {
        success: false,
        errorReason: `Failed to submit transaction: ${error instanceof Error ? error.message : "Unknown error"}`,
        transaction: "",
        network: requirements.network
      };
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ExactAvmSchemeV1
});
