// @algorandfoundation/algokit-utils/packages/crypto/src/hd.mjs
import { BIP32DerivationType, KeyContext, XHDWalletAPI, fromSeed, harden } from "@algorandfoundation/xhd-wallet-api";

//#region packages/crypto/src/hd.ts
const verifyPath = (bip44Path) => {
	if (bip44Path.length !== 5) throw new Error("BIP44 path must have exactly 5 elements");
	if (bip44Path[0] !== harden(44)) throw new Error("BIP44 path must start with 44'");
	if (bip44Path[1] !== harden(283)) throw new Error("BIP44 path must have hardened coin type 283' for Algorand");
};
const getPathComponents = (bip44Path) => {
	return {
		account: bip44Path[2],
		index: bip44Path[4]
	};
};
const xhd = new XHDWalletAPI();
const peikertXHdWalletGenerator = async (seed) => {
	const seedArray = seed ?? new Uint8Array(32);
	if (seed === void 0) crypto.getRandomValues(seedArray);
	const rootKey = fromSeed(Buffer.from(seedArray));
	const rawHdSigner = async (bytesToSign, bip44Path) => {
		verifyPath(bip44Path);
		const { account, index } = getPathComponents(bip44Path);
		return xhd.signAlgoTransaction(rootKey, KeyContext.Address, account, index, bytesToSign, BIP32DerivationType.Peikert);
	};
	return {
		hdRootKey: rootKey,
		rawHdSigner
	};
};
const peikertXHdAccountGenerator = async (rootKey, account, index) => {
	const ed25519Pubkey = await xhd.keyGen(rootKey, KeyContext.Address, account, index, BIP32DerivationType.Peikert);
	const rawEd25519Signer = async (bytesToSign) => {
		return xhd.signAlgoTransaction(rootKey, KeyContext.Address, account, index, bytesToSign, BIP32DerivationType.Peikert);
	};
	return {
		ed25519Pubkey,
		rawEd25519Signer,
		bip44Path: [
			harden(44),
			harden(283),
			harden(account),
			0,
			index
		]
	};
};

//#endregion
export { peikertXHdAccountGenerator, peikertXHdWalletGenerator };
