// @x402-avm/hono/dist/esm/index.mjs
// src/index.ts
import {
  x402HTTPResourceServer,
  x402ResourceServer
} from "@x402-avm/core/server";

// src/adapter.ts
var HonoAdapter = class {
  /**
   * Creates a new HonoAdapter instance.
   *
   * @param c - The Hono context object
   */
  constructor(c) {
    this.c = c;
  }
  /**
   * Gets a header value from the request.
   *
   * @param name - The header name
   * @returns The header value or undefined
   */
  getHeader(name) {
    return this.c.req.header(name);
  }
  /**
   * Gets the HTTP method of the request.
   *
   * @returns The HTTP method
   */
  getMethod() {
    return this.c.req.method;
  }
  /**
   * Gets the path of the request.
   *
   * @returns The request path
   */
  getPath() {
    return this.c.req.path;
  }
  /**
   * Gets the full URL of the request.
   *
   * @returns The full request URL
   */
  getUrl() {
    return this.c.req.url;
  }
  /**
   * Gets the Accept header from the request.
   *
   * @returns The Accept header value or empty string
   */
  getAcceptHeader() {
    return this.c.req.header("Accept") || "";
  }
  /**
   * Gets the User-Agent header from the request.
   *
   * @returns The User-Agent header value or empty string
   */
  getUserAgent() {
    return this.c.req.header("User-Agent") || "";
  }
  /**
   * Gets all query parameters from the request URL.
   *
   * @returns Record of query parameter key-value pairs
   */
  getQueryParams() {
    const query = this.c.req.query();
    const result = {};
    for (const [key, value] of Object.entries(query)) {
      result[key] = value;
    }
    return result;
  }
  /**
   * Gets a specific query parameter by name.
   *
   * @param name - The query parameter name
   * @returns The query parameter value(s) or undefined
   */
  getQueryParam(name) {
    return this.c.req.query(name);
  }
  /**
   * Gets the parsed request body.
   * Requires appropriate body parsing middleware.
   *
   * @returns The parsed request body
   */
  async getBody() {
    try {
      return await this.c.req.json();
    } catch {
      return void 0;
    }
  }
};

// src/index.ts
import { x402ResourceServer as x402ResourceServer2, x402HTTPResourceServer as x402HTTPResourceServer2 } from "@x402-avm/core/server";
import { RouteConfigurationError } from "@x402-avm/core/server";
function checkIfBazaarNeeded(routes) {
  if ("accepts" in routes) {
    return !!(routes.extensions && "bazaar" in routes.extensions);
  }
  return Object.values(routes).some((routeConfig) => {
    return !!(routeConfig.extensions && "bazaar" in routeConfig.extensions);
  });
}
function paymentMiddlewareFromHTTPServer(httpServer, paywallConfig, paywall, syncFacilitatorOnStart = true) {
  if (paywall) {
    httpServer.registerPaywallProvider(paywall);
  }
  let initPromise = syncFacilitatorOnStart ? httpServer.initialize() : null;
  let bazaarPromise = null;
  if (checkIfBazaarNeeded(httpServer.routes) && !httpServer.server.hasExtension("bazaar")) {
    bazaarPromise = import("@x402-avm/extensions/bazaar").then(({ bazaarResourceServerExtension }) => {
      httpServer.server.registerExtension(bazaarResourceServerExtension);
    }).catch((err) => {
      console.error("Failed to load bazaar extension:", err);
    });
  }
  return async (c, next) => {
    const adapter = new HonoAdapter(c);
    const context = {
      adapter,
      path: c.req.path,
      method: c.req.method,
      paymentHeader: adapter.getHeader("payment-signature") || adapter.getHeader("x-payment")
    };
    if (!httpServer.requiresPayment(context)) {
      return next();
    }
    if (initPromise) {
      await initPromise;
      initPromise = null;
    }
    if (bazaarPromise) {
      await bazaarPromise;
      bazaarPromise = null;
    }
    const result = await httpServer.processHTTPRequest(context, paywallConfig);
    switch (result.type) {
      case "no-payment-required":
        return next();
      case "payment-error":
        const { response } = result;
        Object.entries(response.headers).forEach(([key, value]) => {
          c.header(key, value);
        });
        if (response.isHtml) {
          return c.html(response.body, response.status);
        } else {
          return c.json(response.body || {}, response.status);
        }
      case "payment-verified":
        const { paymentPayload, paymentRequirements, declaredExtensions } = result;
        await next();
        let res = c.res;
        if (res.status >= 400) {
          return;
        }
        const responseBody = Buffer.from(await res.clone().arrayBuffer());
        c.res = void 0;
        try {
          const settleResult = await httpServer.processSettlement(
            paymentPayload,
            paymentRequirements,
            declaredExtensions,
            { request: context, responseBody }
          );
          if (!settleResult.success) {
            const { response: response2 } = settleResult;
            const body = response2.isHtml ? String(response2.body ?? "") : JSON.stringify(response2.body ?? {});
            res = new Response(body, {
              status: response2.status,
              headers: response2.headers
            });
          } else {
            Object.entries(settleResult.headers).forEach(([key, value]) => {
              res.headers.set(key, value);
            });
          }
        } catch (error) {
          console.error(error);
          res = c.json({}, 402);
        }
        c.res = res;
        return;
    }
  };
}
function paymentMiddleware(routes, server, paywallConfig, paywall, syncFacilitatorOnStart = true) {
  const httpServer = new x402HTTPResourceServer(server, routes);
  return paymentMiddlewareFromHTTPServer(
    httpServer,
    paywallConfig,
    paywall,
    syncFacilitatorOnStart
  );
}
function paymentMiddlewareFromConfig(routes, facilitatorClients, schemes, paywallConfig, paywall, syncFacilitatorOnStart = true) {
  const ResourceServer = new x402ResourceServer(facilitatorClients);
  if (schemes) {
    schemes.forEach(({ network, server: schemeServer }) => {
      ResourceServer.register(network, schemeServer);
    });
  }
  return paymentMiddleware(routes, ResourceServer, paywallConfig, paywall, syncFacilitatorOnStart);
}
export {
  HonoAdapter,
  RouteConfigurationError,
  paymentMiddleware,
  paymentMiddlewareFromConfig,
  paymentMiddlewareFromHTTPServer,
  x402HTTPResourceServer2 as x402HTTPResourceServer,
  x402ResourceServer2 as x402ResourceServer
};
