// @x402-avm/avm/dist/cjs/exact/v1/client/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/exact/v1/client/index.ts
var client_exports = {};
__export(client_exports, {
  ExactAvmSchemeV1: () => ExactAvmSchemeV1
});
module.exports = __toCommonJS(client_exports);

// src/exact/v1/client/scheme.ts
var import_transact2 = require("@algorandfoundation/algokit-utils/transact");
var import_common3 = require("@algorandfoundation/algokit-utils/common");

// src/utils.ts
var import_algod_client = require("@algorandfoundation/algokit-utils/algod-client");
var import_transact = require("@algorandfoundation/algokit-utils/transact");
var import_common = require("@algorandfoundation/algokit-utils/common");

// src/constants.ts
var ALGORAND_MAINNET_CAIP2 = "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
var ALGORAND_TESTNET_CAIP2 = "algorand:SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
var V1_ALGORAND_MAINNET = "algorand-mainnet";
var V1_ALGORAND_TESTNET = "algorand-testnet";
var V1_TO_CAIP2 = {
  [V1_ALGORAND_MAINNET]: ALGORAND_MAINNET_CAIP2,
  [V1_ALGORAND_TESTNET]: ALGORAND_TESTNET_CAIP2
};
var CAIP2_TO_V1 = {
  [ALGORAND_MAINNET_CAIP2]: V1_ALGORAND_MAINNET,
  [ALGORAND_TESTNET_CAIP2]: V1_ALGORAND_TESTNET
};
var USDC_MAINNET_ASA_ID = "31566704";
var USDC_TESTNET_ASA_ID = "10458941";
var USDC_DECIMALS = 6;
var USDC_CONFIG = {
  [ALGORAND_MAINNET_CAIP2]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [ALGORAND_TESTNET_CAIP2]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  // V1 network mappings
  [V1_ALGORAND_MAINNET]: {
    asaId: USDC_MAINNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  },
  [V1_ALGORAND_TESTNET]: {
    asaId: USDC_TESTNET_ASA_ID,
    name: "USDC",
    decimals: USDC_DECIMALS
  }
};
var FALLBACK_ALGOD_MAINNET = "https://mainnet-api.algonode.cloud";
var FALLBACK_ALGOD_TESTNET = "https://testnet-api.algonode.cloud";
var DEFAULT_ALGOD_MAINNET = typeof process !== "undefined" && process.env?.ALGOD_MAINNET_URL || FALLBACK_ALGOD_MAINNET;
var DEFAULT_ALGOD_TESTNET = typeof process !== "undefined" && process.env?.ALGOD_TESTNET_URL || FALLBACK_ALGOD_TESTNET;
var NETWORK_TO_ALGOD = {
  [ALGORAND_MAINNET_CAIP2]: DEFAULT_ALGOD_MAINNET,
  [ALGORAND_TESTNET_CAIP2]: DEFAULT_ALGOD_TESTNET,
  [V1_ALGORAND_MAINNET]: DEFAULT_ALGOD_MAINNET,
  [V1_ALGORAND_TESTNET]: DEFAULT_ALGOD_TESTNET
};

// src/utils.ts
var import_algod_client2 = require("@algorandfoundation/algokit-utils/algod-client");
var import_common2 = require("@algorandfoundation/algokit-utils/common");
function createAlgodClient(network, customUrl, token = "") {
  const url = customUrl ?? NETWORK_TO_ALGOD[network] ?? DEFAULT_ALGOD_TESTNET;
  return new import_algod_client.AlgodClient({ baseUrl: url, token: token || void 0 });
}
function encodeTransaction(txn) {
  return Buffer.from(txn).toString("base64");
}

// src/exact/v1/client/scheme.ts
var ExactAvmSchemeV1 = class {
  /**
   * Creates a new ExactAvmSchemeV1 instance.
   *
   * @param signer - The AVM signer for client operations
   * @param config - Optional configuration for Algod client
   */
  constructor(signer, config) {
    this.signer = signer;
    this.config = config;
    this.scheme = "exact";
  }
  /**
   * Creates a payment payload for the Exact scheme (V1).
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements
   * @returns Promise resolving to a payment payload
   */
  async createPaymentPayload(x402Version, paymentRequirements) {
    const selectedV1 = paymentRequirements;
    const v1Network = selectedV1.network;
    const caip2Network = V1_TO_CAIP2[v1Network] ?? v1Network;
    const algodClient = this.config?.algodClient ?? createAlgodClient(
      caip2Network,
      this.config?.algodUrl ?? DEFAULT_ALGOD_TESTNET,
      this.config?.algodToken
    );
    const suggestedParams = await algodClient.suggestedParams();
    const assetId = this.getAssetId(selectedV1.asset, caip2Network);
    const assetTransferTxn = new import_transact2.Transaction({
      type: import_transact2.TransactionType.AssetTransfer,
      sender: import_common3.Address.fromString(this.signer.address),
      fee: suggestedParams.fee ?? suggestedParams.minFee ?? BigInt(1e3),
      firstValid: suggestedParams.firstValid,
      lastValid: suggestedParams.lastValid,
      genesisHash: suggestedParams.genesisHash,
      genesisId: suggestedParams.genesisId,
      note: new Uint8Array(Buffer.from(`x402-payment-v${x402Version}`)),
      assetTransfer: {
        receiver: import_common3.Address.fromString(selectedV1.payTo),
        amount: BigInt(selectedV1.maxAmountRequired),
        assetId: BigInt(assetId)
      }
    });
    const encodedTxn = (0, import_transact2.encodeTransactionRaw)(assetTransferTxn);
    const signedTxns = await this.signer.signTransactions([encodedTxn], [0]);
    const signedTxn = signedTxns[0];
    if (!signedTxn) {
      throw new Error("Failed to sign transaction");
    }
    const payload = {
      paymentGroup: [encodeTransaction(signedTxn)],
      paymentIndex: 0
    };
    return {
      x402Version,
      scheme: selectedV1.scheme,
      network: selectedV1.network,
      payload
    };
  }
  /**
   * Gets the asset ID from the requirements or defaults to USDC
   *
   * @param asset - The asset identifier from requirements
   * @param network - The network identifier
   * @returns The resolved ASA ID string
   */
  getAssetId(asset, network) {
    if (/^\d+$/.test(asset)) {
      return asset;
    }
    const usdcConfig = USDC_CONFIG[network];
    if (usdcConfig) {
      return usdcConfig.asaId;
    }
    return asset;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ExactAvmSchemeV1
});
